{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Motif Mapping\n",
    "\n",
    "Using Ciera Martinez's map_motif.py in a generalized jupyter notebook function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "import Bio\n",
    "\n",
    "# For computations\n",
    "import string\n",
    "import random\n",
    "import matplotlib.pyplot as plt; plt.rcdefaults()\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "from IPython.display import Image\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from glob import glob\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython.display import display\n",
    "%matplotlib inline\n",
    "plt.style.use('fivethirtyeight')\n",
    "\n",
    "import os, sys\n",
    "from Bio import motifs\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC, generic_dna, generic_protein"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Processing\n",
    "\n",
    "Reads in the alligned raw DNA sequences and motifs to be pipelined:\n",
    "\n",
    "1. Saves a copy of the sequence without the '$-$' included.\n",
    "\n",
    "2. Casts the undashed version as IUPACUnambiguousDNA().\n",
    "\n",
    "3. Calculates the thresholds to be used in assigning scores for likihood of the motif appearing.\n",
    "\n",
    "4. Takes the indicies that have been scored and maps them to the corresponding raw sequence index, taking into consideration if the strand was positive or negative.\n",
    "\n",
    "5. Returns a df with $\\scriptsize{[\\textbf{position}, \\textbf{score, sequence length}, \\textbf{species}, \\textbf{raw_position index}, \\textbf{strand positioning}, \\textbf{aligned position}]}$.\n",
    "\n",
    "$\\textbf{*calculate_all_TFBS(files, all_motifs) will start the pipeline}$"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following reads in the files from the local directory and filters them according to the file types.\n",
    "\n",
    "Both sequences and motifs are saved as lists. [files has 3540 fasta files & all_motifs has 91 motif files]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Reads through and saves the names of all same files(.fa and .fm respectively) found in the same directory.\n",
    "\n",
    "files = glob('../data/alignments/*0.fa')\n",
    "all_motifs = glob('../data/pwm/*.fm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "def read_records(file):\n",
    "    \"\"\"\n",
    "    file: a single .fa file within the directory.\n",
    "    \n",
    "    Returns the records from a .fa file as a list .\n",
    "    \"\"\"\n",
    "    return list(SeqIO.parse(file, \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "def read_motif(motif):\n",
    "    \"\"\"\n",
    "    motif: a single .fm file within the directory.\n",
    "    \n",
    "    Returns a motif.jaspar.Motif (the motif matrix) from the .fm file.\n",
    "    \"\"\"\n",
    "    return motifs.read(open(motif), \"pfm\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following functions clean the read in alligned sequences, taking out the '-' in them and then casting them as needed. We then feed this cleaned format to create a pssm (PositionSpecificScoringMatrix) with a patser_threshold applied on it to output a score with a corresponding index. Depending on the sign of the position we labeled the DNA strand as positive or negative and then saved all the calculated values as columns in a pandas df."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def raw_sequence_ungap(records):\n",
    "    \"\"\"\n",
    "    records: a list of species within a single .fa file.\n",
    "    \n",
    "    Returns a list of Sequences with an ungapped('-') sequence.\n",
    "    \"\"\"\n",
    "    raw_sequences = []\n",
    "    for record in records:\n",
    "        raw_sequences.append(SeqRecord(record.seq.ungap(\"-\"), id = record.id))\n",
    "    return raw_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def cast_sequence(ungapped_sequence):\n",
    "    \"\"\"\n",
    "    ungapped_sequence: a list with the sequence and id for all the species in a file.\n",
    "    \n",
    "    Returns a list sequences with the type cast as c.\n",
    "    \"\"\"\n",
    "    casted = []\n",
    "    for record in ungapped_sequence:\n",
    "        casted.append(Seq(str(record.seq), IUPAC.IUPACUnambiguousDNA()))\n",
    "    return casted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def calculate_pssm(chosen_motif, pseudocounts=0.0):\n",
    "    \"\"\"\n",
    "    chosen_motif: Chosen motif sequence to be used in creating the pwm\n",
    "    pseudocounts: Default set to 0.0, otherwise added to the counts before calculating position-weight matrix.\n",
    "    \n",
    "    Returns the pssm (PositionSpecificScoringMatrix)\n",
    "    \"\"\"\n",
    "    pwm = chosen_motif.counts.normalize(pseudocounts)\n",
    "    return pwm.log_odds()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def extract_len_id(raw_seq, raw_seq2):\n",
    "    \"\"\"\n",
    "    raw_seq: A list of raw sequences that hasn't yet been cast.\n",
    "    raw_seq2: A list of cast & ungapped sequences.\n",
    "    \n",
    "    Returns a list of dictionary pairs for the species and seq_len of each sequence.\n",
    "    \"\"\"\n",
    "    raw_id = []\n",
    "    record_length = []\n",
    "    for seq in raw_seq:\n",
    "        try:\n",
    "            raw_id.append(seq.id)\n",
    "        except AttributeError:\n",
    "            raw_id.append('n/a')\n",
    "            \n",
    "    for record in raw_seq2:\n",
    "            record_length.append(len(record))\n",
    "    return [{'species': x, 'seq_len': y} for x, y in zip(raw_id, record_length)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below are functions used to find the corresponding positions in the sequence and then mark them as negative or positive strands based on if the value was below or above 0 respectively. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def positions(raw_sequence, cast_sequences, motif, chosen_precision=10**4):\n",
    "    \"\"\"\n",
    "    raw_sequence: A list of the sequences that have been ungapped but not casted.\n",
    "    cast_sequences: A list of sequences that have been ungapped and casted.\n",
    "    motif: the chosen motif file read in using read_motif\n",
    "    precision: default precision set to 10**4 to be used in the pssm distribution.\n",
    "    \n",
    "    Returns a list of positions\n",
    "    \"\"\"\n",
    "    pssm = calculate_pssm(motif)\n",
    "    distribution = pssm.distribution(background=motif.background, precision= chosen_precision)\n",
    "    patser_threshold = distribution.threshold_patser()\n",
    "    \n",
    "    position_list = []\n",
    "    len_and_ids = extract_len_id(raw_sequence, cast_sequences)\n",
    "    for i in range(len(len_and_ids)):\n",
    "        for position, score in pssm.search(cast_sequences[i], threshold= patser_threshold):\n",
    "            pos = {'species': len_and_ids[i].get('species'), 'score':score, \n",
    "                         'position':position, 'seq_len': len_and_ids[i].get('seq_len') }\n",
    "            position_list.append(pos)\n",
    "    return position_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def positive_positions(df):\n",
    "    \"\"\"\n",
    "    df: The df with position, score, seq_len, and species.\n",
    "    \n",
    "    Returns a df where the position arguments are translated into positive ints.\n",
    "    \"\"\"\n",
    "    temp_pos = df[df[\"position\"] >= 0].copy()\n",
    "    temp_neg = df[df[\"position\"] < 0].copy()\n",
    "    temp_pos[\"raw_position\"] = temp_pos[\"position\"]\n",
    "    temp_neg[\"raw_position\"] = temp_neg[\"seq_len\"] + temp_neg[\"position\"]  \n",
    "    temp_together = temp_pos.append(temp_neg).reset_index().sort_values(\"index\")\n",
    "\n",
    "    return temp_together.set_index(\"index\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def define_sign(df):\n",
    "    \"\"\"\n",
    "    df: The df with position, score, seq_len, species, and raw_position.\n",
    "    \n",
    "    Returns the df with a strand sign column appended on.\n",
    "    \"\"\"\n",
    "    df[\"strand\"] = np.where(df['position'] >= 0, 'positive', 'negative')\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def merge_align_df(raw_df, aligned_seq, raw_id_len):\n",
    "    \"\"\"\n",
    "    raw_df: A pandas df that contains the positions, score, seq_len, species, and strand orientation.\n",
    "    aligned_seq: Original inputted sequence with the '-' still included.\n",
    "    raw_id_len: The length of the o\n",
    "    \n",
    "    Returns a df with alignned index appended to raw_df sorted by ['species', 'raw_position'] \n",
    "    with N/A values dropped.\n",
    "    \n",
    "    \"\"\"\n",
    "    remap_list = []\n",
    "    nuc_list = ['A', 'a', 'G', 'g', 'C', 'c', 'T', 't', 'N', 'n']\n",
    "\n",
    "    for i in range(raw_id_len):\n",
    "        counter = 0\n",
    "        for xInd, x in enumerate(aligned_seq[i].seq):    \n",
    "            if x in nuc_list:\n",
    "                remaps = {'raw_position': counter, 'align_position':xInd, 'species':aligned_seq[i].id}\n",
    "                counter += 1\n",
    "                remap_list.append(remaps)\n",
    "            \n",
    "    remap_DF = pd.DataFrame(remap_list)\n",
    "    TFBS_map_DF_all = pd.merge(raw_df, remap_DF, on=['species', 'raw_position'], how='outer')\n",
    "    TFBS_map_DF_all = TFBS_map_DF_all.sort_values(by=['species','align_position'], ascending=[True, True])\n",
    "\n",
    "    return TFBS_map_DF_all.dropna()   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def save_df(TFBS_df, align_file, motif_file):\n",
    "    \"\"\"\n",
    "    TFBS_df: A pandas df with the raw/aligned position, score, and file names.\n",
    "    align_file: Name of read in alignment file.\n",
    "    motif_file: Name of read in motif file.\n",
    "    \n",
    "    Returns a csv copy of the df stored in the current directory\n",
    "    \"\"\"\n",
    "    return TFBS_df.to_csv(motif_file + \".csv\", sep='\\t', na_rep=\"NA\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating a Standard Score Distribution\n",
    "The following 4 cells generate a random sequence, finds the 95 percentile cutoff point in the scores of the random sequence (differs depending on which motif you use), and uses this as the standard cutoff point for your own observed scores."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Randomly generates a sequence\n",
    "def generate_one_seq():\n",
    "    header = \"VT0000|1|MEMB001A|-|\"+str(np.random.randint(1000,9000))\n",
    "    random_list = [random.choice(\"ACGT\") for i in range(5000)]\n",
    "    sequence = \"\".join(random_list) # 1 long string\n",
    "    return header + \"\\n\" + sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Creates a fasta file with randomly generated sequences\n",
    "ofile = open(\"../data/alignments/random_fasta.txt\", \"w\")\n",
    "\n",
    "for i in range(40):\n",
    "    ofile.write(\">\" + generate_one_seq() + \"\\n\")\n",
    "\n",
    "ofile.close()\n",
    "\n",
    "standard_file = glob('../data/alignments/*random_fasta.txt')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def stand_cutoff(file, motifs, motif_to_use):\n",
    "    \"\"\"\n",
    "    files: one fasta file (string)\n",
    "    motifs: a list of motifs\n",
    "    motif_to_use: a number to index into (chose 1 motif you want to use)\n",
    "    Returns the 95th percentile of the scores, and the mean of the scores \n",
    "    \"\"\"\n",
    "    motif = motifs[motif_to_use]\n",
    "    curr_file = read_records(file)\n",
    "    curr_motif = read_motif(motif)\n",
    "    curr_raw = raw_sequence_ungap(curr_file)\n",
    "    curr_cast = cast_sequence(curr_raw)\n",
    "\n",
    "    len_raw = len(extract_len_id(curr_raw, curr_cast))\n",
    "    raw_df = pd.DataFrame(positions(curr_raw, curr_cast, curr_motif))\n",
    "    temp_df = positive_positions(raw_df)\n",
    "    percentile_95 = np.percentile(temp_df[\"score\"], 95)\n",
    "    standard_mean = temp_df[\"score\"].mean()\n",
    "    return (percentile_95, standard_mean)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The following cell prints out:\n",
    "1. the standard 95 percentile \"cutoff\" point for the 2nd motif ('../data/pwm/cad_FlyReg.fm')\n",
    "2. The 50th percentile (mean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3.4494204521179199, 1.0504150713263924)\n"
     ]
    }
   ],
   "source": [
    "print(stand_cutoff(standard_file, all_motifs, 2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Extraction\n",
    "\n",
    "Taking the dfs that come from the \"calculate_all_TFBS\" function we filter out those with a score above the 95th percentile. With the filtered out values we match the species with their respective DNA sequences and then locate and save the given motif within the DNA sequence. Whether the strand is positive or negative is treated individually and the -'s are included in the sequence indexing. This newly extracted information is saved as columns and appended to the df."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following 3 cells provide a convenient way to access the standard \"cutoff\" point when using the 4 motifs printed below. cutoff_dict is used in the filter_95_percentile function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/pwm/bcd_FlyReg.fm',\n",
       " '../data/pwm/cad_FlyReg.fm',\n",
       " '../data/pwm/eve_new6.fm',\n",
       " '../data/pwm/zelda_.fm']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cutoff_dict = {}\n",
    "cutoff_dict[\"bcd_FlyReg.fm\"] = stand_cutoff(standard_file, all_motifs, 0)[0]\n",
    "cutoff_dict[\"cad_FlyReg.fm\"] = stand_cutoff(standard_file, all_motifs, 1)[0]\n",
    "cutoff_dict[\"eve_new6.fm\"] = stand_cutoff(standard_file, all_motifs, 2)[0]\n",
    "cutoff_dict[\"zelda_.fm\"] = stand_cutoff(standard_file, all_motifs, 3)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'bcd_FlyReg.fm': 3.3171534538269043,\n",
       " 'cad_FlyReg.fm': 2.6483244895935059,\n",
       " 'eve_new6.fm': 3.4494204521179199,\n",
       " 'zelda_.fm': 6.6781315803527832}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cutoff_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def filter_95_percentile(TFBS_df, motif_used):\n",
    "    \"\"\"\n",
    "    motif_used: a string that specifies the motif you used (e.g. \"cad_FlyReg.fm\")\n",
    "    TFBS_df: A pandas df with the raw/aligned position, score, and file names.\n",
    "    \n",
    "    returns the TFBS_df with only values above the 95th percentile.\n",
    "    \"\"\"\n",
    "    minimum = cutoff_dict[motif_used]\n",
    "    filtered_df = TFBS_df[TFBS_df[\"score\"] >= minimum]\n",
    "    return filtered_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def calculate_one_dfs_TFBS(file, all_motifs):\n",
    "    \"\"\"\n",
    "    file: A single .fa file from the current directory.\n",
    "    all_motifs: A list of all the .fm files in the current directory.\n",
    "    \n",
    "    Returns a df with the positions, score, and orientation saved into the current directory for all files.\n",
    "    \"\"\"\n",
    "\n",
    "    for motif in all_motifs:\n",
    "        curr_file = read_records(file)\n",
    "        curr_motif = read_motif(motif)\n",
    "        curr_raw = raw_sequence_ungap(curr_file)\n",
    "        curr_cast = cast_sequence(curr_raw)\n",
    "\n",
    "        len_raw = len(extract_len_id(curr_raw, curr_cast))\n",
    "        raw_df = pd.DataFrame(positions(curr_raw, curr_cast, curr_motif))\n",
    "        temp_df = positive_positions(raw_df)\n",
    "        temp_df = define_sign(temp_df)\n",
    "\n",
    "        align_name = re.split(r'_', file)[-1]\n",
    "        final_df = merge_align_df(temp_df, curr_file, len_raw)\n",
    "\n",
    "    return final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>score</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>species</th>\n",
       "      <th>raw_position</th>\n",
       "      <th>strand</th>\n",
       "      <th>align_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>23.0</td>\n",
       "      <td>2.304673</td>\n",
       "      <td>2035.0</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>23.0</td>\n",
       "      <td>positive</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td>-1918.0</td>\n",
       "      <td>2.304673</td>\n",
       "      <td>2035.0</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>117.0</td>\n",
       "      <td>negative</td>\n",
       "      <td>139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>125.0</td>\n",
       "      <td>2.304673</td>\n",
       "      <td>2035.0</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>125.0</td>\n",
       "      <td>positive</td>\n",
       "      <td>154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td>-1879.0</td>\n",
       "      <td>2.304673</td>\n",
       "      <td>2035.0</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>156.0</td>\n",
       "      <td>negative</td>\n",
       "      <td>191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469</th>\n",
       "      <td>-1856.0</td>\n",
       "      <td>2.304673</td>\n",
       "      <td>2035.0</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>179.0</td>\n",
       "      <td>negative</td>\n",
       "      <td>230</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     position     score  seq_len                   species  raw_position  \\\n",
       "465      23.0  2.304673   2035.0  VT0850|1|MEMB002A|-|2035          23.0   \n",
       "466   -1918.0  2.304673   2035.0  VT0850|1|MEMB002A|-|2035         117.0   \n",
       "467     125.0  2.304673   2035.0  VT0850|1|MEMB002A|-|2035         125.0   \n",
       "468   -1879.0  2.304673   2035.0  VT0850|1|MEMB002A|-|2035         156.0   \n",
       "469   -1856.0  2.304673   2035.0  VT0850|1|MEMB002A|-|2035         179.0   \n",
       "\n",
       "       strand  align_position  \n",
       "465  positive              25  \n",
       "466  negative             139  \n",
       "467  positive             154  \n",
       "468  negative             191  \n",
       "469  negative             230  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# EXAMPLE\n",
    "TFBSdf = calculate_one_dfs_TFBS(files[0], all_motifs)\n",
    "TFBSdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# LOOP THROUGH & APPLY TO ALL THE ALIGNMENTS\n",
    "tables = []\n",
    "for i in files[:50]:\n",
    "    tables.append(calculate_one_dfs_TFBS(i, [all_motifs[1]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>species</th>\n",
       "      <th>raw_position</th>\n",
       "      <th>strand</th>\n",
       "      <th>align_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>472</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>255.0</td>\n",
       "      <td>positive</td>\n",
       "      <td>325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>328.0</td>\n",
       "      <td>positive</td>\n",
       "      <td>441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>332.0</td>\n",
       "      <td>negative</td>\n",
       "      <td>445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>336.0</td>\n",
       "      <td>negative</td>\n",
       "      <td>482</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>399.0</td>\n",
       "      <td>positive</td>\n",
       "      <td>584</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        score                   species  raw_position    strand  \\\n",
       "472  6.678132  VT0850|1|MEMB002A|-|2035         255.0  positive   \n",
       "475  6.678132  VT0850|1|MEMB002A|-|2035         328.0  positive   \n",
       "476  6.678132  VT0850|1|MEMB002A|-|2035         332.0  negative   \n",
       "477  6.678132  VT0850|1|MEMB002A|-|2035         336.0  negative   \n",
       "480  6.678132  VT0850|1|MEMB002A|-|2035         399.0  positive   \n",
       "\n",
       "     align_position  \n",
       "472             325  \n",
       "475             441  \n",
       "476             445  \n",
       "477             482  \n",
       "480             584  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# EXAMPLE\n",
    "# Only show the rows with scores above the 95% threshold.\n",
    "# Also removes a few unnecessary columns\n",
    "filtered = filter_95_percentile(TFBSdf, \"cad_FlyReg.fm\")\n",
    "filtered = filtered.drop(['seq_len', 'position'], axis=1)\n",
    "filtered.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# LOOP THROUGH & APPLY TO ALL THE ALIGNMENTS\n",
    "filtered_tables = []\n",
    "for i in tables:\n",
    "    f = filter_95_percentile(i, \"cad_FlyReg.fm\").drop(['seq_len', 'position'], axis=1)\n",
    "    filtered_tables.append(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def map_raw_seq(TFBS_df, seq_file):\n",
    "    \"\"\"\n",
    "    TFBS_df: The TFBS_df with entries only above the 95th percentile.\n",
    "    seq_file: a single .fa file within the directory.\n",
    "    \n",
    "    returns a df with scores in the upper 5th percentile along with the species'\n",
    "    corresponding raw sequence.\n",
    "    \"\"\"\n",
    "    \n",
    "    values = [[record.seq, record.description] for record in seq_file]\n",
    "    sequences = pd.DataFrame(values)\n",
    "    sequences.rename(columns={0 : 'Sequence', 1: 'Description'}, inplace=True)\n",
    "    sequences = sequences.set_index(\"Description\")\n",
    "    mapping = sequences.to_dict(\"index\")\n",
    "    \n",
    "    upper_fifth = filter_95_percentile(TFBS_df, \"cad_FlyReg.fm\")\n",
    "    upper_fifth = upper_fifth[[\"species\", \"raw_position\",'align_position', \"strand\"]]\n",
    "    upper_fifth[\"sequence\"] = upper_fifth[\"species\"].apply(lambda x: str(mapping.get(x).get(\"Sequence\")))\n",
    "    \n",
    "    upper_fifth = upper_fifth.reset_index().drop(\"index\", axis=1)\n",
    "    \n",
    "    return upper_fifth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def positive_index_seq(matched_df, motif):\n",
    "    \"\"\"\n",
    "    matched_df: df with the species matched with their corresponding raw sequence.\n",
    "    motif: a single .fm file within the directory.\n",
    "    \n",
    "    returns the df with appended columns of the DNA sequences/indices starting from the \n",
    "    raw_index from positively oriented scores and then going on until the proper number\n",
    "    of letters have been found.\n",
    "    \"\"\"\n",
    "    motif = str(first_motif.consensus)\n",
    "    motif_len = len(motif)\n",
    "    \n",
    "    extracted_seqs = []\n",
    "    extracted_indices = []\n",
    "    \n",
    "    pos = matched_df[matched_df[\"strand\"] == \"positive\"]\n",
    "    final = pos.copy()\n",
    "    for i in np.arange(pos.shape[0]):\n",
    "        curr_row = pos.iloc[i]\n",
    "        curr_seq = curr_row[\"sequence\"]\n",
    "        index = int(curr_row[\"raw_position\"])\n",
    "        \n",
    "        selected_seq = curr_seq[index:index + motif_len]     \n",
    "        letters = re.findall(r'[^-]', selected_seq)\n",
    "        additional_indicies = 0\n",
    "        \n",
    "        if len(letters) != motif_len: \n",
    "            while len(re.findall(r'[^-]', selected_seq)) != motif_len:\n",
    "                additional_indicies += 1\n",
    "                selected_seq = curr_seq[index:index + motif_len + additional_indicies]\n",
    "                 \n",
    "        selected_index = \"{0}:{1}\".format(index, index + motif_len + additional_indicies)\n",
    "        extracted_seqs.append(selected_seq)\n",
    "        extracted_indices.append(selected_index)\n",
    "        \n",
    "    final[\"extracted_seq\"] = extracted_seqs\n",
    "    final[\"extracted_indices\"] = extracted_indices\n",
    "    return final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def negative_index_seq(matched_df, motif):\n",
    "    \"\"\"\n",
    "    matched_df: df with the species matched with their corresponding raw sequence.\n",
    "    motif: a single .fm file within the directory.\n",
    "    \n",
    "    returns the df with appended columns of the DNA sequences/indices starting from the \n",
    "    raw_index from negatively oriented scores and then going on until the proper number\n",
    "    of letters have been found.\n",
    "    \"\"\"\n",
    "    motif = str(first_motif.consensus)\n",
    "    motif_len = len(motif)\n",
    "    \n",
    "    extracted_seqs = []\n",
    "    extracted_indices = []\n",
    "    \n",
    "    d={'A':'T','T':'A', 'C':'G', 'G':'C'}\n",
    "    \n",
    "    neg = matched_df[matched_df[\"strand\"] == \"negative\"]\n",
    "    final = neg.copy()\n",
    "    for i in np.arange(neg.shape[0]):\n",
    "        curr_row = neg.iloc[i]\n",
    "        curr_seq = curr_row[\"sequence\"]\n",
    "        \n",
    "        index = int(curr_row[\"raw_position\"])\n",
    "        \n",
    "        selected_seq = curr_seq[-(index + motif_len + 1): -(index + 1)]     \n",
    "        letters = re.findall(r'[^-]', selected_seq)\n",
    "        additional_indicies = 0\n",
    "        \n",
    "        if len(letters) != motif_len: \n",
    "            while len(re.findall(r'[^-]', selected_seq)) != motif_len:\n",
    "                additional_indicies += 1\n",
    "                selected_seq = curr_seq[-(index + motif_len + additional_indicies + 1): -(index + 1)]\n",
    "                 \n",
    "        selected_index = \"{0}:{1}\".format(-(index + motif_len + additional_indicies + 1), -(index + 1)) \n",
    "        \n",
    "        swapped_seq = ''.join(d[s] if s in d else s for s in selected_seq)\n",
    "        reversed_seq = swapped_seq[::-1]\n",
    "        extracted_seqs.append(reversed_seq)\n",
    "        extracted_indices.append(selected_index)\n",
    "        \n",
    "    final[\"extracted_seq\"] = extracted_seqs\n",
    "    final[\"extracted_indices\"] = extracted_indices\n",
    "    return final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def find_motif_pipeline(TFBS_df, seq_file, motif):\n",
    "    \"\"\"\n",
    "    TFBS_df: A pandas df with the raw/aligned position, score, and file names.\n",
    "    seq_file: a single .fa file within the directory.\n",
    "    motif: a single .fm file within the directory.\n",
    "    \n",
    "    returns the execution of the pipeline for filtering out the calculated motif locations\n",
    "    and the actual DNA sequence.\n",
    "    \"\"\"\n",
    "    matched = map_raw_seq(TFBS_df, seq_file)\n",
    "    final_p = positive_index_seq(matched, motif)\n",
    "    final_n = negative_index_seq(matched, motif)\n",
    "    final = final_p.reset_index().merge(final_n.reset_index(), how='outer').sort_values('index').set_index('index')\n",
    "    final[\"gap_less_seq\"] = [re.sub(r'-', '', x) for x in final[\"extracted_seq\"]]\n",
    "    return final[[\"species\", \"raw_position\", \"strand\",\"gap_less_seq\", \"extracted_indices\", 'align_position']]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A pipeline for reading in the list of fasta files and motifs and then generating all possible combinations between the two lists to create the specific motif's sequence logo and the corresponding stripplot for locations of the motif against all the species. Currently saves the images into the current directory. \n",
    "\n",
    "Example output for $\\scriptsize\\textbf{Fasta}$: 'align_outlier_rm_with_length_VT0809.fa'\n",
    "$\\scriptsize\\textbf{Motif}$: 'bcd_cell2008.fm'\n",
    "\n",
    "<img src=\"attachment:Screen%20Shot%202018-04-26%20at%209.23.37%20AM.png\" width=\"80%\">\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### *final_df is the scored table of everything \n",
    "#### *indexed_df has the upper 95th percentile / extracted motifs and their locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "def calculate_one_TFBS(file, motif):\n",
    "    \"\"\"\n",
    "    file: A single .fa file in the current directory.\n",
    "    motif: A single .fm file in the current directory.\n",
    "    \n",
    "    Returns a csv with the positions, score, and orientation saved into the current directory for the specified \n",
    "    .fa and .fm files.\n",
    "    \"\"\"\n",
    "    curr_file = read_records(file)\n",
    "    curr_motif = read_motif(motif)\n",
    "    curr_raw = raw_sequence_ungap(curr_file)\n",
    "    curr_cast = cast_sequence(curr_raw)\n",
    "            \n",
    "    len_raw = len(extract_len_id(curr_raw, curr_cast))    \n",
    "    raw_df = pd.DataFrame(positions(curr_raw, curr_cast, curr_motif))\n",
    "    temp_df = positive_positions(raw_df)\n",
    "    temp_df = define_sign(temp_df)\n",
    "            \n",
    "    align_name = re.split(r'_', files[0])[-1]\n",
    "    final_df = merge_align_df(temp_df, curr_file, len_raw)\n",
    "    save_df(final_df, align_name, motif)\n",
    "    \n",
    "    new_path = curr_motif.weblogo(\"{}.png\".format(motif), format= 'PNG', stack_width= 'large')\n",
    "    display(Image(\"{}.png\".format(motif), width= 200, height= 300))\n",
    "\n",
    "    indexed_df = find_motif_pipeline(final_df, curr_file, curr_motif)\n",
    "    testing_pos = indexed_df[indexed_df[\"strand\"] == 'positive']\n",
    "    testing_neg = indexed_df[indexed_df['strand'] == 'negative']\n",
    "\n",
    "    fig = plt.figure(figsize=(20, 15))\n",
    "    plt.subplot(2, 1, 1)\n",
    "    sns.stripplot(x=\"align_position\", y=\"species\", hue=\"strand\", data=testing_pos, dodge=True, linewidth=1)\n",
    "    plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.,title='Orientation', fontsize='x-large')\n",
    "    plt.xlabel('')\n",
    "    plt.title('{} & {} Alignment'.format(align_name, motif))\n",
    "\n",
    "    plt.subplot(2, 1, 2)\n",
    "    sns.stripplot(x=\"align_position\", y=\"species\", hue=\"strand\", data=testing_neg,\n",
    "                  palette=\"Set2\", dodge=True, linewidth=1)\n",
    "    plt.legend(bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.,title='Orientation', fontsize='x-large')\n",
    "    plt.xlabel('')\n",
    "    \n",
    "    #fig.savefig('{} & {}.png'.format(align_name, motif), dpi=fig.dpi, bbox_inches='tight', pad_inches=0.5)\n",
    "    return final_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#calculate_one_TFBS(files[0],all_motifs[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "def calculate_all_TFBS(files, all_motifs):\n",
    "    \"\"\"\n",
    "    files: A list of all the .fa files in the current directory.\n",
    "    all_motifs: A list of all the .fm files in the current directory.\n",
    "    \n",
    "    Returns a csv with the positions, score, and orientation saved into the current directory for all files.\n",
    "    \"\"\"\n",
    "    for file in files:\n",
    "        for motif in all_motifs:\n",
    "            calculate_one_TFBS(file, motif)\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#calculate_all_TFBS(files[:1], all_motifs[:2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Run this cell to generate the plots for all the files with all the possible motif files.\n",
    "#Will save the scored df, motif sequence logo, and file/motif stripplot into the current directory.\n",
    "\n",
    "#calculate_all_TFBS(files, all_motifs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### June 21 Task - Begin reading from here"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtered_tables\n",
    "* filtered_tables is a LIST of TABLES; each table in this list represents ONE alignment file. In every table, 1 row represents a spot on the sequence that we've chosen to label as TFBS, based on our 95% criteria. You can index into filtered_tables to view any one table. For example, alignment 1:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>species</th>\n",
       "      <th>raw_position</th>\n",
       "      <th>strand</th>\n",
       "      <th>align_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10649</th>\n",
       "      <td>2.648324</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>10.0</td>\n",
       "      <td>positive</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10684</th>\n",
       "      <td>2.648324</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>47.0</td>\n",
       "      <td>positive</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10707</th>\n",
       "      <td>2.648324</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>71.0</td>\n",
       "      <td>positive</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10710</th>\n",
       "      <td>2.648324</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>75.0</td>\n",
       "      <td>positive</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10740</th>\n",
       "      <td>2.700792</td>\n",
       "      <td>VT0850|1|MEMB002A|-|2035</td>\n",
       "      <td>106.0</td>\n",
       "      <td>negative</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          score                   species  raw_position    strand  \\\n",
       "10649  2.648324  VT0850|1|MEMB002A|-|2035          10.0  positive   \n",
       "10684  2.648324  VT0850|1|MEMB002A|-|2035          47.0  positive   \n",
       "10707  2.648324  VT0850|1|MEMB002A|-|2035          71.0  positive   \n",
       "10710  2.648324  VT0850|1|MEMB002A|-|2035          75.0  positive   \n",
       "10740  2.700792  VT0850|1|MEMB002A|-|2035         106.0  negative   \n",
       "\n",
       "       align_position  \n",
       "10649              12  \n",
       "10684              56  \n",
       "10707              92  \n",
       "10710              96  \n",
       "10740             128  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_tables[0].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### position_species_tables\n",
    "(See below for example) position_species_tables is also a LIST of TABLES. Again, each element in the list represents ONE alignment file. Each table tells you:\n",
    "\n",
    "1. The alignment position\n",
    "2. A tuple containing all the species that have TFBS's at that position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "position_species_tables = []\n",
    "\n",
    "for f in filtered_tables:\n",
    "    positions_with_TFBS = list(f[\"align_position\"])\n",
    "    combos = {} # Dictionary with position index : list of species that have TFBS at this position\n",
    "    for i in positions_with_TFBS:\n",
    "        combos[i] = list(f[f[\"align_position\"] == i][\"species\"])\n",
    "\n",
    "    d = {'Alignment Positions': list(combos.keys()),\n",
    "         'Species with TFBS at Position': list(combos.values())}\n",
    "    positions_species = pd.DataFrame(data=d)\n",
    "    positions_species[\"Species with TFBS at Position\"] = [tuple(i) for i in positions_species[\"Species with TFBS at Position\"]]\n",
    "\n",
    "    position_species_tables.append(positions_species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Alignment Positions</th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1024</td>\n",
       "      <td>(VT0850|1|MEMB006B|+|2193, VT0850|1|MEMB006C|+...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1025</td>\n",
       "      <td>(VT0850|1|MEMB002A|-|2035, VT0850|1|MEMB002B|-...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3075</td>\n",
       "      <td>(VT0850|1|MEMB006C|+|2323,)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Alignment Positions                      Species with TFBS at Position\n",
       "0                 1024  (VT0850|1|MEMB006B|+|2193, VT0850|1|MEMB006C|+...\n",
       "1                 1025  (VT0850|1|MEMB002A|-|2035, VT0850|1|MEMB002B|-...\n",
       "2                 3075                        (VT0850|1|MEMB006C|+|2323,)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position_species_tables[0].head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### group_by_combo_tables\n",
    "Another LIST of tables. Similar to position_species_tables, except each table provides:\n",
    "\n",
    "1. Index: the ID of each unique combination (starts from 0, 1, 2.....)\n",
    "2. Species with TFBS at Position: a tuple containing a list of species, in other words, 1 combination.\n",
    "3. Count: The frequency at which each combination occurs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "group_by_combo_tables = []\n",
    "\n",
    "for p in position_species_tables:\n",
    "    group_by_combo = p.groupby(\"Species with TFBS at Position\",sort=False).count().reset_index()\n",
    "    group_by_combo.columns = [\"Species with TFBS at Position\", \"Count\"]\n",
    "    group_by_combo_tables.append(group_by_combo)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(VT0850|1|MEMB006B|+|2193, VT0850|1|MEMB006C|+...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(VT0850|1|MEMB002A|-|2035, VT0850|1|MEMB002B|-...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(VT0850|1|MEMB006C|+|2323,)</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Species with TFBS at Position  Count\n",
       "0  (VT0850|1|MEMB006B|+|2193, VT0850|1|MEMB006C|+...      2\n",
       "1  (VT0850|1|MEMB002A|-|2035, VT0850|1|MEMB002B|-...      1\n",
       "2                        (VT0850|1|MEMB006C|+|2323,)     11"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_by_combo_tables[0].head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### counts_at_positions_tables\n",
    "Another LIST of tables. Each table is the merging of the respective tables from position_species and group_by_combo. This one summarizes all the information we've obtained so far."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "counts_at_positions_tables = []\n",
    "num_alignments = len(group_by_combo_tables)\n",
    "\n",
    "for n in range(num_alignments):\n",
    "    counts_at_positions = position_species_tables[n].merge(group_by_combo_tables[n], how=\"left\",sort=False)\n",
    "    counts_at_positions.set_index(\"Alignment Positions\", inplace=True)\n",
    "    counts_at_positions_tables.append(counts_at_positions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alignment Positions</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1024</th>\n",
       "      <td>(VT0850|1|MEMB006B|+|2193, VT0850|1|MEMB006C|+...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1025</th>\n",
       "      <td>(VT0850|1|MEMB002A|-|2035, VT0850|1|MEMB002B|-...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3075</th>\n",
       "      <td>(VT0850|1|MEMB006C|+|2323,)</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Species with TFBS at Position  Count\n",
       "Alignment Positions                                                          \n",
       "1024                 (VT0850|1|MEMB006B|+|2193, VT0850|1|MEMB006C|+...      2\n",
       "1025                 (VT0850|1|MEMB002A|-|2035, VT0850|1|MEMB002B|-...      1\n",
       "3075                                       (VT0850|1|MEMB006C|+|2323,)     11"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_at_positions_tables[0].head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "position_to_count_dictionaries = []\n",
    "\n",
    "for c in counts_at_positions_tables:\n",
    "    position = list(c.index.values)\n",
    "    dictionary = {}\n",
    "    for i in range(len(position)):\n",
    "        if i in position:\n",
    "            dictionary[i] = c[\"Count\"][i]\n",
    "        else:\n",
    "            dictionary[i] = 0\n",
    "    position_to_count_dictionaries.append(dictionary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def freq_at_position(align_position, alignment):\n",
    "    \"\"\"\n",
    "    align_position: integer\n",
    "    alignment: integer\n",
    "    \n",
    "    Returns the frequency of the pattern occurring at the specified alignment position.\n",
    "    Must specify which alignment file is being used.\n",
    "    \"\"\"\n",
    "    return position_to_count_dictionaries[alignment][align_position]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example: this is the frequency of the pattern occuring at position 12 in the first alignment.\n",
    "freq_at_position(12, 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualizations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x132c7a588>"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcAAAAF3CAYAAAAy6pu+AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzs3XdUFOf+BvBnQFDAIIs0BREDKEgQxRKDFRO7EomxoDFq\nYomxxQoa+alXc8FubowtmBixRexKVDS2KAIx9hAMBpFYaFJUBEXY3x9e9rrCLgPMLCjP5xzOYWdm\nZ559Z5YvszvzvkJWVpYSRERE1YxeZQcgIiKqDCyARERULbEAEhFRtcQCSERE1RILIBERVUssgERE\nVC2xABIRUbXEAkhERNVSjbIs/Pfff2PLli1ITExEVlYWlEr1e+gFQcDu3bslDUhERCQH0QUwLCwM\n48aNg56eHhwdHWFmZiZnLiIiIlkJYrtCa9GiBUxNTREWFgYrKyu5cxEREclK9HeA9+7dw8cff8zi\nR0RErwXRBbBp06ZITk6WMwsREZHOiC6ACxcuRGhoKM6fPy9nHiIiIp0Q/R3g4MGDkZCQgBs3bsDF\nxQV2dnbQ01Ovn4IgYNu2bbIEJSIikpLoq0CvXLkCQRBQr149ZGdnIzs7u9gygiBIGo6IiEguos8A\niYiIXifsCYaIiKqlMvUEAwDHjx/HkSNHkJSUBACwt7dHjx494O3tLXk4IiIiuYj+CPTJkycYMWIE\njhw5AgCq+wFTU1MBAD169MDGjRthaGgoU1QiIiLpiP4INDg4GIcPH8bUqVORkJCAuLg4xMXF4ebN\nm5g+fToOHTqERYsWyZmViIhIMqLPAN3d3dGpUyesWrWqxPnjx4/H6dOncfXqVUkDEhERyUH0GWBq\naio8PT01zvf09FR9HEpERFTViS6A9evXx5kzZzTOP3PmDOrXry9JKCIiIrmJLoB+fn7Ys2cPpkyZ\ngvj4eBQUFKCwsBDx8fGYOnUq9u3bhyFDhsiZlYiISDKivwMsKCjAhAkTsH37dgiCoOoGrbCwEEql\nEn5+fli1alWx7tGIiIiqojL3BHP58mVERETgn3/+AQA0aNAA3bp1g4eHhywBiYiI5MCu0IiIqFri\n55VERFQtaewKzdPTE3p6eoiKikKNGjXg6elZ6mgPgiBwvEAiInolaCyArVq1AvC/IY6KHhMREb0O\n+B0gERFVS6K/AwwLC1ONAFGS27dvIywsTJJQREREchNdAMeOHYvo6GiN82NiYjB27FhJQhEREclN\ndAFUKrV/UpqXlwd9ff0KByIiItIFrQPi3rlzB7dv31Y9/vvvv0s8C8zKysLGjRthb28vfUIiIiIZ\naL0IJjg4GIsWLSr19gelUgk9PT2sXLkSw4YNkzwkERGR1LQWwD///BN//vknlEolRo0ahVGjRuGd\nd95RX4EgwNjYGB4eHrCxsZE9MBERkRS0fgTq6uoKV1dXAEBOTg46duwIBwcHXeQiIiKSFe8DJCKi\naknjGeCyZcsgCAKmTJkCQRCwbNmyUlcmCAKmTp0qaUAiIiI5aDwDVCgUEAQBycnJMDQ0hEKhKH1l\ngoCMjAzJQxIREUlNYwEsKCgAANW9fUWPS8N7AYmI6FWg8SPQlwsZCxsREb1OtF4F+qL8/Hzk5ubC\n1NRUNS0zMxOhoaHIzs5G37590bx5c1lCEhERSU30VaCfffYZ/vjjD/z6668AgNzcXLRv3x4JCQkA\nAAMDAxw8eBBt2rSRLy0REZFERPcFevbsWfTq1Uv1OCwsDAkJCdi+fTv++OMPODk5YcmSJbKEJCIi\nkproApieng47OzvV40OHDqF169bo3r076tevj48++giXL1+WJSQREZHURBdAExMTZGdnAwCePXuG\ns2fPonPnzqr5xsbGePjwoeQBiYiI5CD6IhgPDw+EhobC29sbP//8Mx49eoQePXqo5t+8eRNWVlay\nhCQiIpKa6AI4Z84c+Pr6okOHDlAqlejbty88PT1V88PDw/H222/LEpKIiEhqZeoLNDU1FVFRUahT\npw46duyoGiYpKysLoaGh6NChA2+FICKiVwI7wyYiompJ9EegRY4fP44jR44gKSkJAGBvb4+ePXuq\nXRBDRERU1Yk+A3zy5AlGjBiBI0eOAIDqgpfU1FQAQI8ePbBx40YYGhrKFJWIiEg6om+DCA4OxuHD\nhzF16lQkJCQgLi4OcXFxuHnzJqZPn45Dhw5h0aJFcmYlIiKSjOgzQHd3d3Tq1AmrVq0qcf748eNx\n+vRpXL16VdKAREREchB9Bpiamqp228PLPD09VR+HEhERVXWiC2D9+vVx5swZjfPPnDmD+vXrSxKK\niIhIbqILoJ+fH/bs2YMpU6YgPj4eBQUFKCwsRHx8PKZOnYp9+/ZhyJAhcmYlIiKSjOjvAAsKCjBh\nwgRs374dgiBAT+957SwsLIRSqYSfnx9WrVqlmk5ERFSVlflG+EuXLuHo0aP4559/AAANGjRAt27d\n4OHhUeaNh4SE4IcfflCty8XFBdOnT0e3bt1UywQFBWHTpk3IyspCy5YtsXTpUri4uJR5W0RERC+q\n1J5gDh06BENDQzg6OqKwsBBbt27F119/jVOnTqFp06ZYuXIlli9fjtWrV8PJyQmLFi1CVFQUzp8/\nDxMTk8qKTUREr4FSC+DJkyexZs0aJCUlQaFQwNfXF6NHj5YtUKNGjTBv3jwMHz4cLi4uGDt2LKZM\nmQIAyMvLg7OzMxYuXIjhw4fLloGIiF5/WrtC++WXXzBo0CAUFBSgTp06uH79OqKiopCZmYmZM2dK\nGqSwsBB79uzB48eP8fbbbyMxMREpKSnw9vZWLVOrVi14eXkhOjqaBZCIiCpE6xUrK1euRN26dXHq\n1CkkJiYiPj4e7dq1w6pVq/DkyRNJAsTGxsLOzg5WVlaYNm0aNm/eDBcXF6SmpkIQBFhaWqotb2lp\nyfsNiYiowrQWwMuXL+PTTz9Fs2bNAAB169bF3Llz8ejRI9y6dUuSAI0bN8aZM2fwyy+/4NNPP8Vn\nn32GuLg4SdZNRESkidYC+PDhQzRs2FBtWqNGjaBUKpGSkiJJgBo1asDBwQEeHh4IDAyEu7s7Vq9e\nDSsrKyiVSqSlpaktn5aWxpHniYiowkq9ae/l+/qKBsEtLCyUJVBhYSGePHkCBwcHWFtb48SJE6p5\neXl5OHfuHNq2bSvLtomIqPoodTzA7du347ffflM9fvLkCQRBwHfffYfw8HC1ZQVBKNOIEPPnz0e3\nbt1ga2uLR48eISwsDGfPnkVYWBgAYNy4cVi+fDmcnJzg6OiIpUuXonbt2ujfv7/obVRV8fHxcHZ2\nruwYWr0KGQHmlBpzSudVyAi8OjmlVmoBPH78OI4fP15s+svFDyh7AUxJScHYsWORmpoKU1NTuLm5\nYdeuXarBdSdPnoy8vDzMnDlTdSP87t27eQ8gERFVmNYCmJ6eLuvGV69eXeoy/v7+8Pf3lzUHERFV\nP1oLoL6+vq5yEBER6RR7riYiomqJBZCIiKolFkAiIqqWWACJiKhaYgEkIqJqSXQBnDx5Mn7//XeN\n8y9evIjJkydLEoqIiEhuogvgpk2bkJCQoHH+zZs3ERoaKkkoIiIiuUn2EWhGRgZq1qwp1eqIiIhk\npfVG+HPnziEyMlL1+Oeff0ZSUlKx5bKysrBz5064ublJn5CIiEgGWgvgqVOnVH17CoKAvXv3Yu/e\nvSUu6+zsjODgYOkTEhERyUBrAZw4cSI++eQTKJVKuLi4YOnSpejbt6/aMoIgwNjYmB1USyw5GUhN\nff4JtZVVIWxsKjlQNcR9QMD/joPKOAYqc9u6UpnvM60F0MTERFXYLly4ACsrKxY6HUlN1UN4uAEA\noHfvfNjYyDP+ImnGfUDA/46DyjgGKnPbulKZ77NSh0Mq0qhRIzlzEBER6ZToAggA169fx/r163Hp\n0iU8ePCg2KjwgiDg/PnzkgYkIiKSg+jbIKKjo9G5c2fs27cP5ubmuHHjBurVqweFQoGEhATUrFkT\nnp6ecmYlIiKSjOgC+NVXX6F+/fo4f/481q5dCwCYMWMGjh07hoMHD+LOnTsYOnSobEGJiIikJLoA\nXrx4EcOGDYOZmRn09J4/regj0Hbt2mHYsGFYsGCBPCmJiIgkVqaeYBQKBQDA2NgYAJCZmama17hx\nY8TGxkoYjYiISD6iC6C9vT1u3boFAKhZsybs7e1x8uRJ1fyoqChVgSQiIqrqRBdAb29vtV5ghg0b\nhtDQUHzwwQfw9fXF9u3b8eGHH8oSkoiISGqib4OYOnUq+vXrh/z8fBgYGGDatGl49uwZ9u/fDz09\nPUybNg0zZsyQMysREZFkRBdAc3NzmJubqx4LgoCAgAAEBATIEoyIiEhOHBGeiIiqJRZAIiKqllgA\niYioWmIBJCKiaokFkIiIqiUWQCIiqpbKNBwSANy4cQOJiYnIysqCUqksNn/AgAGi17V8+XIcPHgQ\nN27cgKGhIVq1aoW5c+fC1dVVtcznn3+Obdu2qT2vdevWiIiIKGt0IiIiFdEFMDExEePGjUN0dHSJ\nhQ94fm9gWQpgZGQkRo8ejRYtWkCpVOKrr75Cv379EB0dDTMzM9Vy3t7eWL9+vWq7BgYGordBRERU\nEtEF8IsvvsDly5exYMECeHl5qRWo8tq5c6fa43Xr1sHe3h7R0dHo3r27arqhoSEsLCwqvD0iIqIi\nogtgdHQ0Jk+ejPHjx8sW5uHDhygsLCxWXKOiouDs7Iw6deqgXbt2CAwMZEEkIqIKEV0AFQqF7KM9\nBAQEwMPDA23atFFN69q1K3x8fNCwYUMkJSVhwYIF8PHxwalTp/hRKBERlZvoAjhixAiEhYVh1KhR\n0NfXlzzI7NmzERMTg8OHD0MQBNV0X19f1e+urq7w8PCAu7s7jhw5gj59+kieg4iIqgfRBbBx48Y4\nePAgOnTogCFDhsDOzk41MvyLfHx8yhxi1qxZ2Lt3Lw4ePAh7e3uty9rY2KB+/fpISEjQulx8fHyZ\nc+iatozp6XWRkVHzv78/QXz8fV3FUvMqtCMgT0459kF1bk856CJn0XFQ3mOgIhkruu2yqKx9LvZ9\n5uzsLPm2RRfAkSNHqn4PDAwscRlBEJCRkVGmAP7+/ti3bx8OHjwIR0fHUpdPT0/HvXv3YG1trXU5\nORpLSvHx8Voz5ubqwdz8+Ue8Fhb5cHY217isXErLWFXIlVPqfVDd21NquspZdByU5xioaMaKbLss\nKnOfV+bfOtEF8MXBcKUyffp07NixA1u2bIGpqSlSU1MBACYmJjAxMUFOTg6Cg4Ph4+MDa2tr3Lp1\nCwsWLICVlRU//iQiogoRXQA7deok+cY3bNgAQRDw/vvvq0339/eHv78/9PX1ERsbi59++gnZ2dmw\ntrZGx44dsXHjRpiYmEieh4iIqo8y9wSjVCpx5coVJCUlAQDs7e3RrFkztQtXxMrMzNQ6v1atWti1\na1eZ10tERFSaMhXAffv24csvv8Tdu3dVvbIIgoD69evj3//+d7kugCEiIqoMogvg4cOHMXLkSDg6\nOuKrr75CkyZNAADXr1/HDz/8gJEjR2Lr1q1qPbgQERFVVaIL4JIlS9CsWTMcOnQIRkZGquldunTB\nyJEj0b17dyxZsoQFkIiIXgmih0OKjY3FoEGD1IpfkVq1amHw4MH4448/JA1HREQkF9EFsFatWlov\nWsnMzEStWrUkCUVERCQ30QWwU6dOWLt2LWJiYorNO3/+PNauXYvOnTtLmY2IiEg2or8DnDdvHiIj\nI9GjRw+0bNkSTk5OAJ4PkPv777/D0tIS8+bNkysnERGRpEQXQAcHB5w5cwZLly7F0aNHcenSJQCA\nnZ0dxowZg6lTp8LKykq2oERERFIq032AVlZWWLx4sVxZiIiIdEb0d4BERESvE41ngMuWLYMgCJgy\nZQoEQcCyZctKXZkgCJg6daqkAYmIiOSgsQAuXLgQgiBgwoQJMDQ0xMKFC0tdGQsgJScDqanPP1iw\nsiqEjU0lB3pNFbVzedu4os+vql7X1/Wq0NT+VXW/aPwIND09HWlpaTA0NFQ9Lu0nLS1NZ8GpakpN\n1UN4uAHCww1UhZCkV9TO5W3jij6/qnpdX9erQlP7V9X9ovEMUF9fX+tjIiKiV5nocmxpaal1aKK9\ne/fC0tJSklBERERyE10Anz17hsLCwnLPJyIiqkrK9IGstkFvL1y4ADMzswoHIiIi0gWtN8KvW7cO\n69evVz2eM2cOgoKCii2XlZWFzMxMDBw4UPqEREREMtBaABUKBRo1agQASEhIgIWFBWxeuoZVEAQY\nGxujRYsWGDNmjHxJiYiIJKS1AA4cOFB1VtezZ0/4+/tzxAciInotiO4L9NChQ3LmICIi0qkydYYN\nAAUFBbhx4wYePHhQ4lWfb7/9tiTBiIiI5CS6ACqVSvz73//GunXr8OjRI43LZWRkSBKMiIhITqJv\ng/jmm2+wdOlS+Pj4YNWqVVAqlQgMDMSSJUvQpEkTuLu7Y+fOnXJmJSIikozoArhp0yb06dMHq1at\nQo8ePQAAnp6e+PTTT3HixAnk5+cjJiZGtqBERERSEl0A//nnH3h7ez9/kt7zpz19+hQAUKtWLQwe\nPBhbtmyRISIREZH0RBdAMzMz5OXlAQBMTU1hYGCAO3fuqOYbGRnh/v370ickIiKSgegC6OLigj/+\n+OP5k/T00LJlS3z//fdISUnBvXv3sHHjRjg5OckWlIiISEqirwLt378/NmzYgLy8PNSqVQtz5szB\nBx98AFdX1+crqlEDmzdvli0oERGRlESfAX788cc4deoUatWqBQBo164dzp07h/nz52PhwoU4c+YM\nunXrVqaNL1++HF26dIG9vT2cnJwwePBg/Pnnn8WWCwoKgqurK+rVq4c+ffogLi6uTNshIiJ6WYWG\n533zzTcxceJEfP7552jcuHGZnx8ZGYnRo0cjIiICBw4cQI0aNdCvXz9kZWWpllm5ciXWrFmDJUuW\n4MSJE7C0tISvry9ycnIqEp2IiKq5MvcEAwB5eXnIysqCUqksNq9evXqi1/PyfYPr1q2Dvb09oqOj\n0b17dwDA2rVrMWXKFPTp0wcAsGbNGjg7O2Pnzp0YPnx4eeITERGJL4BPnjzBkiVLsHnzZqSmpmpc\nriI9wTx8+BCFhYWqcQUTExORkpKiuv0CeH7LhZeXF6Kjo1kAiYio3EQXwBkzZmDz5s3o3r07vLy8\nUKdOHcnDBAQEwMPDA23atAEApKamQhAEWFpaqi1naWmJ5ORkybdPRETVh+gCuG/fPnz00Uf4z3/+\nI0uQ2bNnIyYmBocPH9Y68rxY8fHxEqSSl7aM6el1kZFR87+/P0F8fOXcY1nWdqys3HLsbzlei1Q5\ni7KVN1dpz38V3j9A8ZwVbZeSVHSdFWlLOV6PJlLsc015tb0Ose8zZ2fnCud7WZk6w27RooXkAQBg\n1qxZ2Lt3Lw4ePAh7e3vVdCsrKyiVSqSlpcHW1lY1PS0tDVZWVlrXKUdjSSk+Pl5rxtxcPZibGwAA\nLCzy4exsrqtoKqVlLEll5C5PTjGkfi1S5izKVt5c2p4vV3tKraScFW2XklRknRVtSzleT0mk2uea\n8mp7HZX5t070VaA9e/bE6dOnJQ/g7++PPXv24MCBA3B0dFSb5+DgAGtra5w4cUI1LS8vD+fOnUPb\ntm0lz0JERNWH6DPAgIAADB8+HFOnTsXHH38MOzs76OvrF1tOoVCI3vj06dOxY8cObNmyBaampqqL\na0xMTGBiYgIAGDduHJYvXw4nJyc4Ojpi6dKlqF27Nvr37y96O0RERC8TXQA9PT0hCAKuXr2KjRs3\nalyuLFeBbtiwAYIg4P3331eb7u/vD39/fwDA5MmTkZeXh5kzZyIrKwstW7bE7t27VQWSiIioPEQX\nwGnTpkm+8czMTFHLvVgQiYiIpCC6AM6ZM0fOHERERDpVoa7QiIiIXlUazwDDwsIAAAMGDFB7XJqi\n5YmIiKoyjQVwzJgxqgtUDA0NMWbMmFJXJggCCyAREb0SNBbACxcuAAAMDQ3VHhMREb0ONBbARo0a\naX1MRET0KivXcEhxcXFISkoCANjb28PFxUXSUERERHIrUwHcv38/5s6di1u3bqlNb9iwIebNm1fs\nhnYiIqKqSnQBDAsLw5gxY/Dmm2/i//7v/1T9dt64cQOhoaEYOXIk1q1bx4tgiIjolSC6AC5evBgt\nWrRAeHg4jIyM1OZ99tln6NmzJxYvXswCSERErwTRN8L/888/GDRoULHiBwBGRkYYPHgwbt++LWk4\nIiIiuYgugE2aNNE6CntKSsorMYYYERERUIYCOH/+fPzwww/Yv39/sXn79u3DDz/8gAULFkgajoiI\nSC4avwMcPHhwsWmWlpYYMWIEbGxsVPcFJiYm4t69e3BycsLatWvRqVMn+dISERFJRGMBvHLlCgRB\nKDa9Xr16AJ4Xvhen5eTk4OrVq9InJCIikoHGAhgbG6vLHERERDrF4ZCIiKhaKnNXaPn5+bh79y6y\nsrKgVCqLzW/evLkkwYiIiOQkugBmZWUhMDAQYWFhePr0abH5SqUSgiAgIyND0oBERERyEF0Ax48f\nj0OHDuH9999Hy5YtYWpqKmcuIiIiWYkugMePH8eYMWMQHBwsZx4iIiKdEH0RjLm5uaoDbCIioled\n6AI4fPhw7N69G4WFhXLmISIi0gnRH4HOnDkTT58+RefOnTFo0CDY2tpCT694/fTx8ZE0IBERkRxE\nF8C7d+/i9OnTuHr1qsYeX3gVKBERvSpEF8AJEybg0qVLmDRpElq2bIk6derImYuIiEhWogtgVFQU\nJk+ejC+//FLOPPSKe/AASEkRYGJSvJMEoqooORlITX3+dY6VVSFsbCo50GssNxe4ckWvyrSz6Itg\nLC0tYW5uLmcWeg08eiQgPl4fOTnFO1InqopSU/UQHm6A8HADVSEkeWRmClWqnUWnGD9+PEJDQ/Ho\n0SM58xAREemE6I9Anz59Cn19fXh6eqJfv36ws7MrdhWoIAgYP358mQJERkbim2++weXLl3Hv3j2s\nXr0afn5+qvmff/45tm3bpvac1q1bIyIiokzbISIiepHoAhgYGKj6/bvvvitxmfIUwJycHLi5ucHP\nzw/jxo0rcRlvb2+sX79e1fm2gYFBmbZBRET0MtEF8MKFC7IE6Nq1K7p27Qrg+dleSQwNDWFhYSHL\n9omIqHoSXQAbNWokZw6toqKi4OzsjDp16qBdu3YIDAxkQSQiogop83iA2dnZOH36NJKSkgAA9vb2\n6Nixo2z3BXbt2hU+Pj5o2LAhkpKSsGDBAvj4+ODUqVP8KJSIiMqtTAVw1apVCAoKQm5urtpguMbG\nxpg9e3aZv/8Tw9fXV/W7q6srPDw84O7ujiNHjqBPnz4anxcfHy95Fqlpy5ieXhcZGTX/+/sTxMff\n11UsNWVtx5ycBsjNVeLx42dIT8/SWW459rcc+0CqnEXZypurtOe/Cu8foHjO8rRLafu5om1dkbas\n6LbLQop9rilv0fScHH1kZBSozRf7PnN2dq5wvpeJLoCbN29GYGAg2rVrh3HjxqFJkyYAgOvXr2Pt\n2rUIDAyEmZkZhg4dKnnIF9nY2KB+/fpISEjQupwcjSWl+Ph4rRlzc/Vgbv78DNfCIh/Ozrq/B7O0\njCX5+299GBnVgrFxASwsaugkd3lyiiH1PpAyZ1G28ubS9ny52lNqJeUsT7uUtp8r0tYVbcuK7mex\npNrnmvIWTTcxeQZz8xpq8yvzb53oArh69Wp06NAB+/btgyD87yZnJycn9OzZE++//z6+/fZb2Qtg\neno67t27B2tra1m3Q0RErzfRN8InJCSgb9++asVPtRI9Pfj4+JR6VlaSnJwcXL16FVeuXEFhYSFu\n376Nq1ev4vbt28jJyUFgYCB+++03JCUl4ddff8WQIUNgZWWl9eNPIiKi0ogugKampqoLX0py69Yt\nmJqaljnAxYsX0bFjR3Tu3Bl5eXkICgpCp06dEBQUBH19fcTGxmLo0KFo3bo1xo8fj8aNGyMiIgIm\nJiZl3hYREVER0R+BduvWDevXr0fz5s3Rv39/tXm7d+9GSEgIBg4cWOYA7du3R2Zmpsb5u3btKvM6\niYiISiO6AM6bNw+//fYbRo8ejTlz5uDNN98E8Pyj0ZSUFDRu3Bhz586VLSgREZGURBdACwsLnDx5\nEhs2bMDRo0fxzz//AAAaN26MiRMnYuTIkTAyMpItKBERkZTKdB+gkZERJkyYgAkTJsiVh4iISCdK\nvQhm3759OHbsmNZljh07hn379kkWioiISG5aC2B4eDhGjhwpakUjR47EkSNHJAlFREQkN60FcMuW\nLWjbti3ee+89rSt577330K5dO2zcuFHKbERERLLRWgDPnz+P7t27i1pR165d8fvvv0sSioiISG5a\nC2BmZqboYYfq1q2LrKwsSUIRERHJTWsBNDMzQ3JysqgVpaSkyDYkEhERkdS0FkBPT0/RV3fu27cP\nLVq0kCQUERGR3LQWwOHDh+Pq1auYM2eO2vh/L1IqlQgMDMTVq1cxYsQIOTLSayY5GbhyRQ8iP1x4\nrRW1RVVuj1chI1F5aL0RvlevXhg6dCi+/fZbHD9+HAMHDkTTpk1Ru3ZtPHr0CLGxsdixYwfi4uIw\nePBg9OrVS1e56RWWmqqH8HAD9O6dDxubwsqOU6mK2gJAlW2PVyEjUXmU2hPMN998A2dnZ6xYsQLz\n589XGw5JqVTC1NQUgYGB+OKLL2QNSkREJKVSC6AgCJg8eTLGjBmDyMhIxMXF4eHDh3jjjTfQpEkT\neHl5wdjYWBdZiYiIJCO6L1AjIyO8++67ePfdd+XMQ0REpBOiB8QlIiJ6nbAAEhFRtcQCSERE1VKZ\nxgMkIpJDcvLz2y2srAphY1PZaai64BkgEVW6onsNU1P5J4l0p0JHW0pKCi5duoTs7Gyp8hAREemE\n1gJ44cIFLFu2DBkZGWrTU1NT0a9fP7i6uqJLly5wdnbGggULZA1KREQkJa0FMCQkBFu3boW5ubna\n9AkTJuDUqVNo06YNxo4dCycnJ6xYsQI//fSTrGGJiIikovUimPPnzxcbDT4xMRFHjx6Ft7c3du/e\nDQB4+vTZcVM/AAAgAElEQVQpvL298eOPP2LQoEHypSUiIpKI1jPA5ORkODs7q02LiIiAIAj45JNP\nVNMMDQ3x4YcfIjY2Vp6UREREEivzRTDR0dEAgHfeeUdtupWVFXJzc6VJRUREJDOtBbBRo0aIiYlR\nPX727Bl+/fVXvPnmm6hbt67ashkZGbCwsJAnJdFr5MXx9R48qOw0RNWX1u8A/fz8EBgYCBcXF7Rt\n2xZhYWFIT0/HqFGjii0bGRkJR0dH2YISvS5eHF+vRYtnlZyGqPrSWgA//fRTHD9+HAsWLIAgCFAq\nlfDy8sLEiRPVlrt9+zaOHTuGwMBAWcMSERFJRetHoAYGBtixYwciIiKwdu1aHDp0COHh4TAyMlJb\n7smTJ1i3bh38/PzKHCAyMhJ+fn5o2rQpFAoFtm3bVmyZoKAguLq6ol69eujTpw/i4uLKvB0iIqIX\niboIpnXr1hg4cCDatm1b4nxHR0d88MEHsLS0LHOAnJwcuLm5ITg4uMSBdVeuXIk1a9ZgyZIlOHHi\nBCwtLeHr64ucnJwyb4uIiKhIubtCy8zMxIkTJ3DgwAHcvHmz3AG6du2KOXPmwMfHB4IgFJu/du1a\nTJkyBX369IGLiwvWrFmDR48eYefOneXeJhERUakF8Ntvv4W3tze6du2KH374AQCwbt06vPXWW+jf\nvz+GDx+Oli1bYty4cSgsLJQ0XGJiIlJSUuDt7a2aVqtWLXh5ealuxyAiIioPrRfBbNu2DXPmzEHD\nhg1hbm6OadOm4f79+/jqq6/QrVs3dOjQAc+ePcORI0fw008/wdXVFZMmTZIsXGpqKgRBKPbRqqWl\nJZKTkyXbDhERVT9aC2BISAi8vLywf/9+6OvrY+nSpQgKCsKAAQOwfv161XJffPEF3n//fWzbtk3S\nAlgR8fHxlR2hVNoypqfXRUZGzf/+/gTx8fd1FUtNWdsxJ6cBcnOVePz4GdLTs0rMXfTapHxdcuxv\nOfZBfHy82npzcvSRkVFQ5m1UtA1Le/6L7ampHR49egOZmYYAAIXiKWrXfljmHGLzaPLyfi/Pekrb\nzxVt64ocm3K8VzSR4j2kKW/R9KLj/cX5Yt9nL/dKJgWtBfDvv//G7Nmzoa+vDwD44IMP8NVXX6FX\nr17Flu3Vqxfmzp0raTgrKysolUqkpaXB1tZWNT0tLQ1WVlZanytHY0kpPj5ea8bcXD2Ymz+/V8zC\nIh/OzuYal5VLaRlL8vff+jAyqgVj4wJYWNQoMXfRa5PqdZUnpxhS74OinC+u18TkGczNa5R5GxVt\nQ23Pf7k9NbXDlSt6iI5+Pr1373w4O5f/K5DyvJ6S9nt51lPafq5IW1f02JT6vaKJVO8hTXmLphcd\n7y/Or8y/dVq/A8zOzoZCoVA9NjMzAwC1aUUUCgWePn0qaTgHBwdYW1vjxIkTqml5eXk4d+6cxitS\niYiIxNB6BqgLOTk5SEhIgFKpRGFhIW7fvo2rV69CoVDAzs4O48aNw/Lly+Hk5ARHR0csXboUtWvX\nRv/+/Ss7OhERvcJKLYAXL15EzZrPP5999OgRBEFAVFRUsVHgL1y4UK4AFy9eRN++fVW3QAQFBSEo\nKAh+fn749ttvMXnyZOTl5WHmzJnIyspCy5YtsXv3bpiYmJRre0RERICIArhmzRqsWbNGbVpwcHCJ\ny5Z0H19p2rdvj8zMTK3L+Pv7w9/fv8zrJiIi0kRrAdy7d6+uchAREemU1gLYqVMnXeUgIiLSKa1X\ngV67dq3Yd31EVP2YmSlRs6YSd+8KeLkPCjOzkqdXJUVjMFbljKR7Wgtgx44dERERoXr86NEjjBo1\niqMxEFUzubnAL78Y4PBhA6Sm6hWbV9L0qqRoDMaqnJF0T+vRoFQq1R4/ffoUu3btQkpKiqyhiIiI\n5MZ/h4iIqFpiASQiomqp1AJY0r195bnfj4iIqCop9Ub4efPmYfny5QCgGu9v0qRJJY7eLggCzp49\nK3FEIiIi6WktgG+//Xaxsz12Qk1ERK8DrQXw8OHDuspBRK8gMzMl8vOVePhQwIMH8m8vOfn5LQ1W\nVuUfeomoiNbvACdPnozff/9dV1mI6BWTmwvEx+vjxg09PHok/7UBvJ+PpKT1KNq0aRMSEhJ0lYWI\niEhn+G8UERFVSyyARERULZV6G0RSUhIuXbokeoXNmzevUCAiIiJdKLUAfvXVV/jqq69KXZFSqYQg\nCMjIyJAkGBERkZxKLYDDhg1Dy5YtdZGFiIhIZ0otgO3bt8eAAQN0kYWIXlEuLgXIzn4+JqCNTWWn\nIRKHF8EQUYUplUBEBO/Po1cLj1YiIqqWtBbAtm3bwtLSUldZiIiIdEbrd4CHDh3SVQ4iIiKd4keg\nRERULbEAEhFRtcQCSERE1ZLG7wDj4uJga2uLN954Q5d5iF5JRePUAYCVVaFO7oUzM1Pi7l0BgJ7O\ntkn0OtF4Bujl5aU2IK6vry9Onz6tk1BEr5qicep0OVZdbi5w+LBut0n0OtH4rjEyMkJubq7q8cmT\nJ5GSkqKTUC8KDg6GQqFQ+3FxcdF5DiIier1o/Aj0rbfewjfffINnz56pPgaNiYkpdYVydJvWuHFj\nhIeHQ6lUAgD09fUl3wYREVUvGgvgokWLMGLECEybNg0AIAgCQkJCEBISonFlgiDIUgD19fVhYWEh\n+XqJiKj60lgAmzdvjgsXLuD27dtIS0vDe++9h1mzZqFLly66zAcAuHXrFlxdXWFoaIhWrVohMDAQ\nDg4OOs9BRESvD609wejp6cHe3h729vYYMGAAvL290apVK11lAwC0bt0aq1evhrOzM9LS0rBkyRJ0\n794d0dHRMDMz02kWIiJ6fZQ6HFKR9evXqz3OysoCANmL0Lvvvqv2uHXr1vDw8MDWrVvx+eefy7pt\nIiJ6fYkugACQmJiIhQsX4ujRo3j48CEA4I033kD37t0xe/ZsnXwsaWxsDBcXFyQkJGhdLj4+XvYs\nFaUtY3p6XWRk1Pzv708QH39fV7HUlLUdc3IaIDdXicePnyE9PavE3EWvTcrXJcf+Lss+ELtsfHy8\n2rI5OfrIyCgQtY2Xt/f4sSkeP9ZDRkZumduytH3wYnsWbSs311BtvxZNf/q0JgoK9JCfX4j09PRy\n7VOxx8SLyzVoUHy/a1qPtvWXtu8qerxW5NiU472iiRTvodLav+h4f3G+2PeOs7NzhfO9THQBvH79\nOrp164bHjx+jW7duaNKkiWr6nj17cPToURw+fFg1XS55eXmIj49Hx44dtS4nR2NJKT4+XmvG3Fw9\nmJsbAAAsLPLh7Gyuq2gqpWUsyd9/68PIqBaMjQtgYVGjxNxFr02q11WenGKUZR+IWbYo54vLmpg8\ng7l5DVHbeHl7xsY1YGysB3NzozK3pbZ98HJ7Fm3LyMhAbb8WTTc01MPzC7RrwMLColz7VOwx8eJy\nwP1i+13TerStv7R9V5HjtaLHptTvFU2keg+V1v5Fx/uL8yvzb53oAjh37lzUqlULx44dK9ZQN27c\nQO/evTF//nxs3bpV0oCBgYHo0aMH7OzsVN8BPn78GH5+fpJuh4iIqhfRBTAyMhITJ04s8b8EJycn\nfPrpp1i1apWk4QDg7t27GD16NO7fvw8LCwu0atUKx44dg52dneTbIiKi6kN0ASwoKECtWrU0zjcy\nMkJBQYEkoV60YcMGyddJREQkugPBZs2aITQ0FNnZ2cXmPXjwAKGhofDw8JA0HBERkVxEnwEGBATg\nww8/RJs2bTB06FA4OjoCeP7937Zt25CRkYGlS5fKFpSIiEhKogtgp06dEBYWhtmzZ2PFihVq89zc\n3LBu3bpSr8wkIiKqKsp0H2Dnzp0RGRmJO3fu4J9//gEANGjQALa2trKEI6LikpOBu3cFPH1a2Uno\nVVEZ41W+CspUAIvY2tqy6BFVktRUPcTG6uPJk8pOQq+KovEqAaB373zY2BRWcqKqgaNoEhFRtcQC\nSERE1RILIBERVUssgEREVC2xABIRUbUkugC6u7tj4cKF+Ouvv+TMQ0REpBOiC6Cbmxv+85//oG3b\ntvD29sa6deuQnp4uZzYieoUlJwNXrughObmyk1ARMzMl7t4VuE/+S3QB3L59O65fv47FixfD0NAQ\nAQEBcHV1xaBBg7B371484U1JRPSConvPim7ApsqXmwscPsx9UqRMraBQKDBq1CgcOXIEFy9exPTp\n05GQkIBPPvkEzs7OmDRpEs6ePStXViIiIsmU+98ABwcH+Pv748CBA+jXrx8ePnyI0NBQ9O3bFx4e\nHli/fj0KC9nbABERVU3l6gotJycHBw4cwI4dO3D69GkIgoBu3bphyJAhMDAwwI8//oiAgADExsZi\n5cqVUmcmIiKqMNEFsLCwEMePH8eOHTvw888/IycnB2+99Rb+9a9/YeDAgbCwsFAt26tXL/zrX/9C\nSEgICyAREVVJogugi4sL0tPTYWlpiREjRsDPzw9ubm4al3dzc8PDhw8lCUlERCQ10QWwffv28PPz\nw7vvvgs9vdK/Ouzfvz/69etXoXBERERyEV0Av//++zKvXF9fv8zPoaqrssYUe3m7wPPHHNeMdK0y\n3gNF9+4BPOalJvoq0CNHjiAgIEDj/FmzZuHo0aOShKKqqei+Ll3f2/Xydnl/GVWWyngPFN27x2Ne\neqJbc+XKlcjOztY4/+HDh7zghYiIXhmiC2BsbCw8PT01zm/evDliY2MlCUVERCQ30QXw2bNnyMvL\n0zg/NzeX3aEREdErQ3QBdHV1xc8//1ziPKVSifDwcDRp0kSyYERERHISXQDHjh2LqKgoDB8+HH/8\n8QeePXuGZ8+e4dq1axg+fDhiYmIwZswYObMSERFJRvRtEAMGDMCNGzewdOlSHDhwQHUvYGFhIQRB\nwLRp0+Dn5ydbUCIiIimVqS/QWbNmYeDAgdi/fz8SExMBAI0aNULfvn3h6OgoRz6qADnvWcrNfT7W\nW3nXW3RvU26uHvLzAVNT3d5bWBkePXoDV67o4cGD/03LzwdSUgQ0bFiIu3cF3L2rDweHAri4lG3d\nRfvD0LAQT59W/H6x5GTg1i09ZGYK0NdXljkHANXrfHHa67yPi95vYl5jWZaVipmZEvn5Sjx7ppvt\nvQrK3Bm2o6MjpkyZIkcWkljRPUsA0Lt3PmxspBudIzNTwPHjNcq93qJ7m+ztC5GUpAdra6XkGaua\nzExDREcboEWL//0FevxYQHy8PhSKQkRH6yMmxgATJ+bBxaWgjOt+vj9atHiGixfLv1+KpKbq4ezZ\nGvjlFwN8/LH4i9uKcgBQvc4Xp73O+7jo/SbmNZZlWank5gI3b+ojN1fQyfZeBeW6qzIvLw/Jycm4\nd+9esR+5hISEwMPDAzY2NujcuTPOnTsn27aIiOj1J/oM8MmTJ1iyZAk2b96M1NRUjctlZGRIEuxF\nu3fvxqxZs7B8+XK0bdsW3333HQYMGIDo6GjY2tpKvj0iInr9iS6AM2bMwObNm9G9e3d4eXmhTp06\ncuZSs3r1anz00UcYNmwYAGDx4sX45Zdf8P333yMwMFBnOYiI6PUhugDu27cPH330Ef7zn//ImaeY\n/Px8XLp0CRMnTlSb3qVLF0RHR+s0CxERvT5EfweoVCrRokULObOU6P79+ygoKICVlZXadEtLS60f\nxRIREWkjZGVlibrGeezYsXj69Cl++OEHuTOpSU5OVvVC884776imL168GDt37kRMTIxO8xAR0etB\n9BlgQEAA/v77b0ydOhWXLl1Ceno6MjMzi/1IrW7dutDX1y92tpeWllbsrJCIiEgs0WeACoUCgiBA\nqVRCEDTfRyLHVaDvvfce3N3dsWLFCtW0Vq1aoV+/fpgzZ47k2yMiotef6Itgpk2bJmcOrcaPH4/P\nPvsMLVq0QNu2bbFhwwakpKRgxIgRlZaJiIhebaLPACvb999/j6+//hopKSlwdXVFUFAQ2rZtW9mx\niIjoFVWuApifn4/MzEwoFAoYGBjIkYuIiEhWZeoKLSoqCr1790b9+vXh4uKCyMhIAM9vVfjggw9w\n8uRJOTKK0rt3bygUCtWPubk5Ro0apbZMVlYWxowZA3t7e9jb22Ps2LHIzs7Wac6q1qVbcHCwWrsp\nFAq4vNQTc1BQEFxdXVGvXj306dMHcXFxsmaKjIyEn58fmjZtCoVCgW3bthVbprRMT58+xYwZM+Do\n6AhbW1v4+fnh7t27Os35+eefF2vbbt266TTn8uXL0aVLF9jb28PJyQmDBw/Gn3/+WWy5ym5PMTmr\nQnuGhISgXbt2qr8h3bp1Q0REhNoyld2WYnJWhbZ82fLly6FQKDBz5ky16XK2p+gCGBkZCR8fH6Sl\npWHEiBFQKv934li3bl3k5+fjxx9/FLs6yQmCgI8++gjx8fH466+/cP36dbWLZgBg1KhRuHbtGvbs\n2YPdu3fjypUr+Oyzz3SWsahLt+nTp+PXX39FmzZtMGDAANy5c0dnGUrSuHFjVbv99ddfqn9sAGDl\nypVYs2YNlixZghMnTsDS0hK+vr7IycmRLU9OTg7c3NwQHBwMY2PjYvPFZAoICEB4eDi+//57HDp0\nCA8fPsSgQYPUjlu5cwKAt7e3Wtvu2LFDbb7cOSMjIzF69GhERETgwIEDqFGjBvr164esrCzVMlWh\nPcXkBCq/PW1tbfGvf/0Lp0+fxsmTJ9GxY0cMHToUsbGxAKpGW4rJCVR+W77ot99+w48//oi33npL\nbbrc7Sn6I9CePXsiLy8PERERePDgAZycnLB371506tQJALBo0SJs2bIFV65cKcvrlkyfPn3QtGlT\nLF68uMT5f/31F95++21ERESgdevWAJ6f0fbs2RPnz5/XyXBOJV3N2rJlS/Tr16/SunQLDg7G/v37\n1Yrei1xcXDB27FjVCCB5eXlwdnbGwoULMXz4cNnz2dnZYcmSJWpjTZaWqej4XLNmDfr37w8AuHPn\nDtzd3bFr1y54e3vrJOfnn3+OjIwMbN++vcTnVEbOnJwc2NvbY+vWrejevTuAqtmeJeWsiu0JPB8S\nbt68eRg+fHiVbMuSclaltszOzkbnzp3xzTffIDg4WO3vuNztKfoM8PLlyxg8eDAMDAxKvA2iXr16\nld4zy+7du+Ho6Ih33nkHgYGBePTokWpeTEwM3njjDVXxA4C2bdvCxMREJ12qFXXp1rlzZ7XpVaFL\nt1u3bsHV1RUeHh749NNPVWM9JiYmIiUlRe0gqlWrFry8vCots5hMFy9exLNnz9SWsbW1RZMmTXSe\nOyoqCs7OzmjVqhUmT56M9PR01bxLly7pPOfDhw9RWFgIMzMzAFW3PV/OWaQqtWdhYSF27dqFx48f\n4+23366ybflyziJVpS2/+OIL+Pr6on379mrTddGeom+DqFGjBgoKNI9RdvfuXdSuXVvs6iQ3cOBA\nNGjQADY2NoiLi8O8efMQGxuLXbt2AQBSU1NRt27dYs+zsLDQSeHW1qXbqVOnZN++Jq1bt8bq1avh\n7OyMtLQ0LFmyBD169EBUVBRSU1MhCAIsLS3VnmNpaYnk5ORKySsmU1paGvT19WFubl5sGV3+k9a1\na1f4+PigYcOGSEpKwoIFC+Dj44NTp07BwMAAqampOs8ZEBAADw8PtGnTBkDVbc+XcwJVpz1jY2PR\nrVs35OXloXbt2ti8eTNcXFwQExNTpdpSU06g6rTljz/+iMTERGzYsKHYPF0cm6ILYOvWrbF//358\n/vnnxebl5ORgy5YtaNeundjVibJw4UIsW7ZM43xBEHDgwAG0a9cOH3/8sWq6q6srHBwc0KVLF1y5\ncgXNmjWTNNfr5N1331V73Lp1a3h4eGDr1q1o1apVJaV6Pfj6+qp+LzrDdnd3x5EjR9CnTx+d55k9\nezZiYmJw+PBhrZ1ZVDZNOatKezZu3BhnzpxBdnY29u/fj88++wzh4eE6275YmnK6uLhUiba8ceMG\nFixYgCNHjkBPr1xD01aY6K3OmjULFy9exMCBA3H06FEAwLVr17Bp0yZ07twZ9+/fx4wZMyQNN378\nePz2228af2JiYtCyZcsSn9u8eXPo6+sjISEBAGBlZYX79+8XWy49PV0nXaq9Kl26GRsbw8XFBQkJ\nCbCysoJSqURaWpraMpWZWUwmKysrFBQUFOuVqLLb2sbGBvXr11c7JnWVc9asWdizZw8OHDgAe3t7\n1fSq1p6acpakstqzRo0acHBwgIeHBwIDA+Hu7o7Vq1dXubbUlLMkldGWMTExyMjIwNtvvw0LCwtY\nWFjg7NmzCAkJgaWlJczNzWVvT9EFsFWrVtixYwdu3LihunJyzpw5mDx5MvLz8/HTTz8Vu4KnohQK\nBZycnLT+1KpVq8TnXrt2DQUFBbC2tgYAtGnTBo8ePcJvv/2mWiY6OrrY5+JyMTAwQPPmzYvdKnLi\nxIkqdUN/Xl4e4uPjYWNjAwcHB1hbW+PEiRNq88+dO1dpmcVkat68OWrUqKG2zJ07d3D9+vVKbev0\n9HTcu3dPdUzqKqe/v7+qqLx8sVdVak9tOUtSWe35ssLCQjx58qRKtaW2nCWpjLbs06cPIiMjcebM\nGdVPixYt8OGHH+LMmTNwcnKSvT31AwIC5okN7ODggDFjxqBHjx5o3749+vTpg0mTJmH+/PlwcHAQ\n/8ollpiYiPXr18PExARPnz5FdHQ0pkyZggYNGuDLL7+EIAioW7cuzp8/j7CwMDRr1gx37tzBlClT\n0KpVK4wePVonOd944w0EBQXB2toaRkZGWLx4MaKiorBq1SqYmprqJMPLAgMDUbNmTSiVSty4cQMz\nZszAzZs3sWLFCpiamqKgoAArVqyAk5MTCgoK8OWXXyI1NRUrVqyAoaGhLJlycnJw/fp1pKSkIDQ0\nFG5ubjA1NUV+fr6oTDVr1kRycjJCQkLg5uaG7OxsTJ06FWZmZpg3b55kH/9py6mvr48FCxbgjTfe\nQEFBAa5cuYLJkyejsLAQS5Ys0VnO6dOn46effsLGjRtha2uLnJwc1SXkRfuvKrRnaTlzcnKqRHvO\nnz9f9X65c+cOVq9ejZ07d2L+/Plo1KhRlWjL0nJaWVlVibasWbOm6syv6CcsLAwNGjRQXU0td3uK\n/g6wiCAIaN68OZo3b16+Vy0DAwMDnDp1CuvWrUNOTg5sbW3RvXt3zJw5U60BQkJCMHPmTNXlsr16\n9dJ424QcfH19kZmZiWXLlqm6dAsLC4OdnZ3OMrzs7t27GD16NO7fvw8LCwu0atUKx44dU2WaPHky\n8vLyMHPmTGRlZaFly5bYvXs3TExMZMt08eJF9O3bV7XvgoKCEBQUBD8/P3z77beiMgUHB6NGjRr4\n5JNPkJeXh06dOmHdunWSfvelLeeyZcsQGxuLn376CdnZ2bC2tkbHjh2xceNGnebcsGEDBEHA+++/\nrzbd398f/v7+AMTt48rOqa+vXyXaMyUlBWPHjkVqaipMTU3h5uaGXbt2qa7urgptWVrOvLy8KtGW\nJXl53XK3p+j7AMVe+qqLjxOJiIgqqszDIZVGjuGQiIiIpCb6I9C9e/cWm1ZQUICkpCRs3LgRenp6\nldabCRERUVlJMhxSQUEBevTogS5dumDWrFlS5CIiIpKVJHcf6uvro3///ti0aZMUqyMiIpKdZLff\nZ2dnF+u5nYiIqKoS/R3gvXv3SpyenZ2NyMhIfPPNN7wClIiIXhmiC2DTpk01XgWqVCrh6emJlStX\nShaMiIhITqIvginp+z1BEGBmZoZGjRpJ3g2aXNzd3dGxY0d8++23AICkpCR4eHhg9erVamO5UdWw\nZcsWTJgwAVeuXEGDBg1KXV6hUCAgIEB1k7eunDlzBn379sXBgwcl7xSeKl/v3r2hp6eHAwcOqKZV\n1rFG0hH9HeDHH39c7GfYsGHo27dvlSl+oaGhUCgUakOovKyks9iq3DO+nGJiYhAcHIwHDx6IWn7r\n1q1QKBSqHwsLC7i5uWH8+PEaPyKvKEEQiu2fnTt3Ys2aNaKX1xU5tnv06FH4+PigSZMmqFevHtzd\n3TFkyBDs3LlT8m1VV6dPn4ZCoUDDhg3x9OnTEpfR9HejOv7tuH79OoKDg/HPP/+U6/kFBQWoX78+\n9u/fDwA4fPgwrK2tkZ+fX2zZZcuWYejQoWjatCkUCgWmTZtWoewvK3NXaFVZWFgYGjZsiBs3buDS\npUuiumuzt7dHcnIyDAwMdJCwaomOjsbixYsxdOhQ0X2RCoKAWbNmwcHBAXl5eYiOjsa2bdsQGRmJ\nc+fOaeycvLz8/Pzw4YcfqvU7GhYWhri4OIwbN67Y8snJyahRQ/eHdfv27ZGcnCxp/6irVq1CYGAg\n2rZti0mTJuGNN97ArVu3EBkZiU2bNuHDDz+UbFvV2Y4dO2Bvb4/bt2/j0KFDxbpj06SyjrXKFhcX\nh0WLFqFDhw6iPpV52bVr15CXl6c6Ufn999/h7u5e4t/ghQsXwsrKCp6enrKMQSp67704fpRYgiBg\n9+7dZX5eedy9exdnz57Fhg0bMGfOHPz000+i+yuVq1Pn11WXLl1Uw1ANGzYMZmZmWL16NX7++Wd8\n8MEHkm5LEIQy7Z/K3JdSbrugoABLlixBx44dsW/fvmLzXxy9m8rvyZMn2L9/P2bMmIHw8HDs2LFD\ndAGszn83KnLme+HCBdSrVw82NjYAnhdAT0/PEpe9fPmyalgshUJR7m1qIvoj0NzcXNy8eRMnT57E\n77//jtTUVKSmpuL333/HyZMnkZiYiLy8PLWf3NxcyQNrEhYWBhMTE/Ts2RO+vr7Ys2cPlMrSv95M\nSkqCQqHAtm3b1Kb/+uuv6Ny5M2xsbODp6YmNGzciKCio2E4oOi0PDw+Hl5cXrK2t8c477+CXX35R\nW67ouX/99RfGjBkDe3t7ODo6YsGCBQCeF/ChQ4fC3t4ejRs3xjfffFMs69OnTxEcHIyWLVvC2toa\nTZs2xezZs4u1s5hMwcHB+L//+z8AQLNmzaBQKGBubo6zZ8+W2mYv69ixI5RKJW7duqXWriNHjsSb\nbxaV3CgAABatSURBVL6JevXqoUuXLiUOGhoSEgIvLy/Y2trC3t4eHTp0wMaNG1Xzt2zZAoVCofq4\npU+fPoiIiFDtt6LcL772RYsWqW1DTJYzZ85AoVBg165dWLZsGdzc3GBjY4P3338fN2/eLLUNip7/\nYvuNGzcONjY2uHfvHoYMGQI7Ozs4OTkhMDCw1GPz/v37ePDggcYrqy0sLIpNW7duHby8vGBjYwNn\nZ2dMnDixxK4JlyxZAjc3N9SvXx8+Pj6Ii4tDs2bNMH78eNUyJR3rQPH9UeT48ePo3bs37OzsYGdn\nhw8//BBXr15VW6as7fHdd9+hQ4cOqFevHhwdHeHr64uoqCi1ZcLCwtClSxfUq1cPDg4OGDFihNpx\nWJqff/4ZOTk5+OCDD9C/f38cO3ZM9O1cJR1r165dQ69evVCvXj24ublh6dKlqq9mXmwzd3d3DBgw\nAFFRUXj33XdhY2OD5s2bY/v27WrrK/ra4cyZM5g5cyacnJzQsGFDTJo0Cfn5+Xjw4AHGjRsHBwcH\nODg4YM6cOSVmFXNsiMm0detWjBgxAsDz92LR++/lv58vy8nJQUZGBu7fv4+oqCi4ubmpHl+4cAGN\nGzdGRkZGsbYvbUzIihJ9Bjh//nwMHjwYK1euxJAhQ1Snq/n5+diyZQvmz5+PdevWaf3+TU47duxA\nz549UbNmTfTv3x+rVq3CiRMn0KVLlzKv6/LlyxgwYACsra0xe/Zs1X/jmvpDjY6OxuHDh/HJJ5+g\ndu3aWLduHYYPH45r167BzMwMwP/+Y/r000/RpEkTzJ8/HxEREVixYgXMzMywZcsWtGvXDvPnz8eO\nHTswd+5cNG/eHB06dFBtZ+jQoTh37hyGDx+OJk2a4Pr16wgJCcH169exa9euMmXy8fHB33//jV27\ndiE4OFhVRJo0aVLm9ioaRLNoHenp6ejWrRseP36MsWPHom7dutixYweGDRuGkJAQ1Vnipk2bMGPG\nDPj6+mLs2LHIz89HXFwcYmJiVG+yl79nmTFjBh48eIB79+4hKCio1EIiNkuRr7/+GjVq1MDEiRPx\n4MEDfP311xgzZoxqEGhtXj42BEGAUqlE//790apVKyxcuBAnT57Et99+izfffBMjR47UuC5LS0sY\nGRnhyJEjGDduXKn//U6ZMgVbtmzBkCFDMHbsWNy+fRvr1q3DhQsXcOLECdXZysKFC7Fs2TJ0794d\nXbt2xdWrV9G/f/9i331p+n6rpOlhYWEYO3YsunTpgrlz5+LJkyf48ccf0bt3bxw/fhxOTk5lbo9J\nkyYhNDQUXbt2xdChQ6FUKhETE4PIyEjVOG8rVqzAggUL4Ovri48++ghZWVn47rvv0LNnT5w5c0bt\nHyNNduzYgTZt2sDW1hb9+vXDrFmzsHfvXtXxVxb37t1D3759oaenh6lTp8LY2BihoaGoUaNGicdG\nYmIiRowYgY8++ghDhgzB5s2bMX78eLRo0aLY+3DWrFmwtrbGrFmzcP78eWzevBl16tTBxYsXUa9e\nPcydOxcRERFYvXo1XF1dMXToUNVzxR4bYjJ5eXlh7NixWL9+PaZPn47GjRsDQKl/92fMmKFWJAVB\nUI35KAgCZs6ciRkzZsDe3h6XL18uc9uXW1ZWllLMT+vWrZXjx4/XOH/8+PHKNm3aiFqX1D9nzpxR\nCoKgDAsLU01zcnJSDho0qNiy9vb2yqFDh6oeX7lyRSkIgnLNmjWqaT169FAaGxsr4+LiVNMuXbqk\nNDAwUOrp6amtTxAEZc2aNZWXLl1STTt79qxSEATl0qVLVdMCAgKUgiAohw8frpqWkZGhtLOzU+rr\n6yvnzp2rmn7r1i2lsbGxWv7vvvtOqa+vrzx06JDa9kNCQpR6enrKvXv3ljnTggULlHp6esqrV6+K\naufVq1cr9fT0lHv27FEmJCQoY2Njld//f3t3H9TE0ccB/HuHaBQBIQK+YAWJ74wgL8WgEkUUoS2W\nMoWqlY6C1Dqg1KJiW9ROX3xBoWOlMoLgS60lGHlTrFhAsbYWdUSHmVYxiFCdogVRUKSW7POHT244\n8sIFebG6nxn+yGbv7pfLspvb291LSyNisZiYmJhw52v58uWEZVmSn5/PbfvXX3+RcePGkaFDh5K6\nujrS0NBAXn/9dTJhwgRBx2wbo6+vLxk5cqTW/AzDkHXr1nGvhcZy9OhRwjAMGTduHJfW0NBAtmzZ\nQliWJefOndMb59GjRwnLsuTYsWNc2oIFCwjLsuTTTz/l5XVyciIuLi4dnu9169YRlmWJiYkJmTVr\nFvn4449JYWGhRr4TJ04QhmHI7t27tabv2LGDNDQ0EKVSSfr160f8/Px4+dasWUMYhuH9X8TGxmqU\ndW3fx+3bt4mFhQUJDQ3l5bt58yaxsrIiwcHBBp+PvLw8wjAMiYiI0HluysvLibGxsca+ysrKiEgk\nIjExMR2e3xs3bpC+ffuS7du3c2ne3t5EKpVq5J02bRqZPn263rIWERFBWJYlp0+f5tKqqqqIpaWl\nRhl+5ZVXCMuy5Mcff+TSlEolEYlEZMWKFbzzzTAMmTVrFu/YHh4ehGVZsnjxYi5NXZ+0jV9o2TAk\npn379mmU9Y7+SktLSU5ODklPTycMw5DExESSk5NDli5dSoYMGUJyc3NJTk4OKSgo0LkPhmFIWFiY\n4GMK+RPcBVpeXq73obcjR47U6PLoKXK5HJaWlryrvaCgIOTn5xvcDatSqVBSUgI/Pz+ujxp4+jBg\nHx8frdt4eXnxzs3EiRNhamqKqqoqXj6GYbBo0SLuNcuycHZ2BiEE7777Lpdubm4OiUTC68rJzs6G\nRCLB2LFjUV9fz/1JpVIAT7tsOxOToQgheOutt+Dg4ICJEyciPDwcNjY2+OGHH7jzdfLkSTg5OcHT\n05PbTiQSISwsDLW1tdwvPDMzM9y+fRuXLl16ppj0ERqL2vz582FkZMS9lkqlIIQ803kLDQ3lvZZK\npYL2t3btWiQnJ8PR0RElJSXYvHkzfHx84ObmhgsXLnD5srKyYGpqCm9vb17ZkEgksLa25spGcXEx\nnjx5gvDwcN5xli1b1unPVlRUhPv37yMoKIh37H///RdSqVSjXAIdn4/c3FxusJUuubm5aG1tRWBg\nIO+4pqammDBhgtbjtqdQKEAI4d3zCwoKQmlpaadGOBYVFcHV1RVOTk5c2qBBg/D2229rzS+RSHhP\nLReLxZBIJFrrjbZXdAC4e/Bt6w11fdK23hBaNgyNyVBjxoyBTCYDy7IwMTFBaGgoZDIZGhoaMH36\ndHh5eUEmk/V4D6LgLlAbGxtkZ2djyZIlvAoCeHrDPisrCzY2Nl0eYEcIIThy5AimTp2K6upqLt3F\nxQVNTU04duyYQaPl7t69i+bmZtjb22u8N2rUKK3bDB8+XCNt0KBBWu8ltH/4rZmZGYyNjWFlZaWR\n3nagg1KpREVFBddt0BbDMLh7926nYzIEwzDYunUrRo8eDZFIBFtbW41j1dTUICAgQGPbMWPGgBCC\n6upquLi4IDo6GiUlJfD29oadnR1mzpyJwMBAXrfvsxIai1r7z6Luwu7seTM2Noa1tbXGPoXuLzg4\nGMHBwXj06BHKysqgUCiwd+9ehISE4Pz587C0tERlZSUaGxsxevRoje3blo0///wTgGY5trS05D6n\noSorK0EIwZtvvqn12O3rCiHno6qqCtbW1nq7fdXHdXNz03pcfT/W1TIzM+Hk5ITGxkY0NjYCePpD\nkWEYZGZmYtWqVR3uo62amhqtgzl01RvaHoRtSL0BaJZXMzMz3vZCy0ZnYhLq4cOHaGlpASEERUVF\ncHJywv3790EIwdmzZ/HBBx+gvr4eLMt2uhx2luAGMCoqCjExMZgzZw7ee+897kutrKzE3r17UVZW\nhvj4+G4LVJczZ87g1q1buH37Nm+SKvD0C5bL5d0+XLz9P7matvtT2vKyrPYL8bbbq1QqjBs3Dps3\nb9a636FDh3Y6JkNNnjyZ+wX6LMaMGYMLFy6goKAARUVFKCgoQHp6OsLDw3ulLAFdf950fbeGGjBg\nADw9PeHp6QmxWIxt27bh5MmTCAkJgUqlglgsRlpamtY4O1Op6Brl19raynutUqnAMAx27drF6zHR\npavOh/q4CoVC6z779++vd/uqqiqUlpaCYRiNRquzDaChnrXe0JXevt4wpGx0R72h7/4fAGzYsAHr\n16/v+ft/MKABDAsLA8uy+OKLL7By5UruH4QQAgsLC8THxyMsLKzbAtUlIyMDgwcPRkJCgsaXVFhY\niEOHDqGurg5isVjQ/qysrCASibSO/FMqlV0Sc2fY29vj8uXL8PLy6rJ9dtck3hEjRqCiokIj/dq1\na2AYhjeySyQSISAgAAEBAVCpVFi2bBn27NmDjz76SGeFakjchsTyX+Hq6gpCCDcvyt7eHqdOnYKb\nmxsGDBigczv1nC2lUsnr4dA2+k5dMT548IA3R7RtL4v62IQQiMViyGSyZ/tg/2dnZ4fCwkLU19fr\nHMiijn/48OHcQAxDZGRkoE+fPti9e7dGpX/lyhUkJCTgypUrmDRpkuB9jhgx4rmsN4SUDUMYWm9E\nR0cjJCQEDx48QGhoKOLj4yGRSHD8+HFkZmYiPT0dhJAOf7R0B4N+ji1evBh//PEHjh07huTkZCQn\nJyM/Px9Xr17tlcavpaUFeXl5mDNnDt544w2uIlX/RUZG4smTJxojJPVhWRYymQzHjx/nrW5SWVmp\nMbWhJwUGBqK2thZ79uzReO+ff/5BU1OTwfs0MTEB0PnuPV18fX1x+fJl3pD1lpYWpKWlwcbGhpuf\nee/ePd52LMtiwoQJAJ4usq4vbn3vdyaW501zczN+++03re8VFBSAYRiuWyswMBCtra3YunWrRl6V\nSsV9vzNmzECfPn2QmprKy6NtVR11w/bLL79waQ8fPtQYpu/t7Q1zc3Ns375d60oedXV1HXxSTfPm\nzQMhBJs2bdKZRz3aUttnBqB1+kdbhw8fhoeHBwIDAzXqjZUrV8LY2BhyudyguL29vXHx4kWUlZVx\naffu3evVVXuElg1DmJiYgBAieFv1/b8BAwbA2NgYCxcuhEwmw+PHj+Hu7t5r9/+ATqwEY2xszBtQ\n0Jvy8/PR2NgIPz8/re+PHj0aDg4OkMvliIiIELzf2NhYFBcXw9fXF2FhYWhtbUVqairGjx+P8vLy\nrgrfICEhIcjJycHq1atx9uxZTJkyBYQQVFRUIDs7G/v27TN4DcrJkyeDEILPPvuMW21FJpPpvVoW\n0hUSHR0NhUKB4OBgREREYPDgwcjIyMC1a9eQmprKdVkFBgbCysoKU6ZMgbW1NSorK5GSkgJHR0e9\n0zGcnZ2RlZWF2NhYuLm5gWVZnRPwhcbSFbqie1mtubkZc+fOhYuLC3x8fDBixAg0NTWhuLgYBQUF\nePXVVzF37lwAgKenJ8LDw7Fjxw6Ul5fD29sb/fr1g1KpRG5uLj755BPMnz8fYrEYUVFRSExMREhI\nCGbPno3y8nL89NNPGvMKvb29YWtri8jISKxYsQIMw+D777+HlZUVbt26xeUzNTVFYmIiIiIi4OXl\nhaCgIFhbW6OmpgaFhYUYP348t+6uUNOmTcOCBQuwZ88e3Lhxgxt8dv78eTg6OuLDDz+EnZ0dNmzY\ngA0bNqC6uhqvvfYazM3NcfPmTW5BBl1rdF68eBHXr1/XOQ3F1NQUU6dOxZEjR/D5558LvuJZuXIl\n5HI5goKC8P7773PTIGxtbdHQ0NDpHpdnKVdCy4YhJk2aBCMjIyQmJqKhoQH9+/eHq6srRo4cqXe7\nc+fOwdHRkVstqrS0VOcAIbWMjAzU1NRw56CsrAzbtm0DALzzzjta71kawqAGsKGhAbt27cKZM2dQ\nV1eHnTt3wt3dHfX19UhPT0dAQIDWm63dRS6XQyQSYebMmTrz+Pv7Y+fOnaisrMSoUaO0zmNq/9rZ\n2RmHDx9GXFwcNm3ahGHDhmHt2rW4fv06rl+/rrHts64TqCtf23SGYXDw4EF8++23OHToEI4fPw6R\nSAQ7OzssXboUEydONDgmZ2dnbNy4EampqYiMjIRKpUJeXp7ehlTIZxo8eDBOnDiBjRs3Ii0tDc3N\nzRg/fjwOHDgAf39/Lt+SJUuQmZmJ5ORkNDY2YsiQIVi0aBFiYmL07j88PBy///47MjMzkZKSwo1M\n1fYZhcai77M9y/fY2X2am5vjm2++QUFBATIyMnDnzh1ucEdsbCyioqJ4jXd8fDycnZ2Rnp6OL7/8\nEkZGRrC1tUVQUBCv2zwuLg4ikQjp6en4+eef4ebmBoVCoVER9enTBwcPHkRMTAy++uorWFtbY/ny\n5TAzM0NkZCQvb2BgIIYOHYqEhAQkJSWhpaUFQ4YMgYeHh0YjI/R8JCUlwdHREQcOHMDGjRsxcOBA\nODk58cpmVFQUJBIJkpKSsH37dqhUKgwbNgwymUzroBw1uVwOhmG4HxDa+Pv749SpUzh9+jRmzJih\nNcb2ZW348OE4evQo1q5di8TERIjFYixZsgQmJiZYt24db5lAffVDR/VTR9rnF1o2hMZkZWWFHTt2\nICEhAdHR0WhtbUVSUlKHDeCFCxfg7u4OAGhsbMS1a9c6vOo7cOAA1wvBMAwuXbrEjRqXSqXP3AAK\nfhpETU0N/Pz8cPfuXYwdOxbl5eXIysri+v3d3d0xY8aMXhu80BMWLlyIq1ev8oagU9SLYNKkSZg+\nfbrBV2tUx2JjY7F//37cunXrpVw8+3kmuP9n/fr1ePLkCX799Vety4z5+/ujpKSkywPsLY8fP+a9\nViqVOHnyZJcO0aco6sXSvt6or6+HXC6HVCqljd9zSHAXaHFxMaKiojBq1CitN5jt7Ox49wb+65yd\nnbFgwQLY2dmhuroaaWlpEIlEWLFiRW+HRlHUc2r27NmYNm0axo4di9raWnz33XdoamrC6tWrezs0\nSgvBDWBLS4vetfUaGxu7dEBBb/Px8YFCocCdO3fQt29feHh4IC4uTusEeYr6r3tZn23X1Xx9fZGT\nk4P9+/eDYRg4OTkhKSmJt7oK9fwQfA9w5syZkEgkSElJQX19PRwcHJCdnc3dA5w7dy5YlkV+fn63\nBkxRFEVRXUHwJduyZcugUCjw9ddfc08QJ4SgsrISERERKC0t5T1OhaIoiqKeZ4KvAIGnw2m3bNkC\nQghUKhVYlgUhBAzDIC4uDtHR0d0ZK0VRFEV1GYMaQAC4efMmcnNzoVQqoVKpYG9vj3nz5ulc8JWi\nKIqinkeCGsDHjx8jLy8PEokEkydP7om4KIqiKKpbCboHKBKJEBkZ2eMrdVMURVFUdxE8CMbBwQG1\ntbXdGQtFURRF9RjBDeCqVauQmpqKq1evdmc8FEVRFNUjBE+ELy0thVgsxtSpUyGVSmFvb89b3BV4\nOpl2y5YtXR4kRVEURXU1waNALSwsOt4Zw3T4HC6KoiiKeh4IvgL8+++/uzMOiqIoiupRghtAIyOj\n7oyDoiiKonqU3kEwa9asQVlZGS/t0aNHXfrka4qiKIrqDXobwJSUFFRUVHCv6+vrYWtr+0I994+i\nKIp6ORn8/CJ69UdRFEW9CF6cB/hRFEVRlAFoA0hRFEW9lDocBVpVVYWLFy8CAPccwIqKCgwcOFBr\nfldX1y4Mj6IoiqK6h96J8BYWFmAYhpemfv5fe+p0OhGeoiiK+i/QewWYlJTUU3FQFEVRVI8y+IG4\nFEVRFPUioINgKIqiqJcSbQApiqKolxJtACmKoqiXEm0AKYqiqJcSbQApiqKolxJtACmKoqiX0v8A\nY6vdWOLX4dIAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x132c44780>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pos = list(position_to_count_dictionaries[0].keys())\n",
    "y_pos = np.arange(len(pos))\n",
    "freqs = list(position_to_count_dictionaries[0].values())\n",
    " \n",
    "plt.bar(y_pos, freqs, align='center', alpha=0.5, width = 3)\n",
    "plt.xlabel(\"Alignment Position in Sequence Alignment #1\")\n",
    "plt.ylabel(\"Frequency of TFBS Combination at Position\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ID Dictionary\n",
    "Use the function ID_dictionary (defined below) when you have an ID and want to view the pattern associated with that ID."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# For convenience, here are dictionaries of IDs that map to a combination.\n",
    "\n",
    "all_combo_ids = []\n",
    "\n",
    "for g in group_by_combo_tables:\n",
    "    combo_ids = {}\n",
    "    for i in g.index.values:\n",
    "        combo = g[\"Species with TFBS at Position\"][i]\n",
    "        combo_ids[i] = combo\n",
    "    all_combo_ids.append(combo_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def ID_dictionary(alignment, ID):\n",
    "    \"\"\"\n",
    "    alignment: integer specifying which alignment to analyze\n",
    "    ID: integer; ID linked to the unique TFBS combination\n",
    "    \"\"\"\n",
    "    return all_combo_ids[alignment][ID]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('VT10600|0|MEMB002D|-|719',\n",
       " 'VT10600|0|MEMB002E|-|729',\n",
       " 'VT10600|0|MEMB003A|+|713',\n",
       " 'VT10600|0|MEMB003B|+|702',\n",
       " 'VT10600|0|MEMB003D|-|714',\n",
       " 'VT10600|0|MEMB006B|+|689',\n",
       " 'VT10600|0|MEMB007C|+|679',\n",
       " 'VT10600|0|dkik|+|726')"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example: from alignment 2, ID #11 represents this combination of species.\n",
    "ID_dictionary(2, 11)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TLDR:\n",
    "#### These functions incorporate everything done above.\n",
    "1. Use **frequent_patterns** function to find the top few occurring TFBS patterns in your specified alignment file.\n",
    "2. Use **most_frequent_pattern** function to find the top occuring TFBS pattern in your specified alignment file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(VT0850|1|MEMB002F|+|2216,)</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>(VT0850|1|MEMB002A|-|2035, VT0850|1|MEMB002B|-...</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>(VT0850|1|MEMB007D|+|2300,)</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>(VT0850|1|dkik|-|2904,)</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Species with TFBS at Position  Count\n",
       "4                         (VT0850|1|MEMB002F|+|2216,)     16\n",
       "8   (VT0850|1|MEMB002A|-|2035, VT0850|1|MEMB002B|-...     28\n",
       "9                         (VT0850|1|MEMB007D|+|2300,)     23\n",
       "12                            (VT0850|1|dkik|-|2904,)     23"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def frequent_patterns(alignment_num, percentile=80):\n",
    "    \"\"\"\n",
    "    alignment_num: integer specifying which alignment to analyze\n",
    "    percentile: (integer) limit frequencies to only above this percentile\n",
    "    \n",
    "    Returns:\n",
    "    A table, like those in group_by_combo_tables, except w/only the largest counts\n",
    "    \"\"\"\n",
    "    counts = list(counts_at_positions_tables[alignment_num][\"Count\"])\n",
    "    table = group_by_combo_tables[alignment_num]\n",
    "    return table[table[\"Count\"] >= np.percentile(counts, percentile)]\n",
    "\n",
    "frequent_patterns(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>(VT0870|0|MEMB002F|+|2269,)</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Species with TFBS at Position  Count\n",
       "35   (VT0870|0|MEMB002F|+|2269,)     31"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def most_frequent_pattern(alignment_num):\n",
    "    \"\"\"\n",
    "    alignment_num: integer specifying which alignment to analyze\n",
    "    \n",
    "    Returns:\n",
    "    A single-row table with the TFBS combo with the LARGEST count in the sequence\n",
    "    \"\"\"\n",
    "    max_count = max(list(counts_at_positions_tables[alignment_num][\"Count\"]))\n",
    "    table = group_by_combo_tables[alignment_num]\n",
    "    return table[table[\"Count\"] == max_count]\n",
    "\n",
    "most_frequent_pattern(1) # <- In the second alignment, the most frequent combo occurs 52 times."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### More Analysis..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following cells join the tables together from all 50 alignments into **frequent_all_alignments**. There are no duplicate species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/JoanneChen/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "      <th>Count</th>\n",
       "      <th>Alignment File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>(VT19240|1|MEMB003F|-|4104,)</td>\n",
       "      <td>109</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(VT19020|1|MEMB003A|-|4766, VT19020|1|MEMB003D...</td>\n",
       "      <td>94</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(VT10600|0|MEMB002F|-|2671,)</td>\n",
       "      <td>84</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(VT18830|1|MEMB002F|+|3408,)</td>\n",
       "      <td>82</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>(VT14350|0|MEMB002F|+|3647,)</td>\n",
       "      <td>72</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>(VT19750|1|MEMB002A|+|2196, VT19750|1|MEMB002B...</td>\n",
       "      <td>67</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(VT19600|0|MEMB002A|-|2450, VT19600|0|MEMB002B...</td>\n",
       "      <td>63</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(VT19350|0|MEMB002A|+|2674, VT19350|0|MEMB002B...</td>\n",
       "      <td>56</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(VT17000|1|MEMB002A|-|2389, VT17000|1|MEMB002B...</td>\n",
       "      <td>56</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(VT19920|1|MEMB003B|+|3304,)</td>\n",
       "      <td>56</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(VT14350|0|MEMB002B|+|3460, VT14350|0|MEMB004A...</td>\n",
       "      <td>53</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>(VT19060|0|MEMB002A|+|2590, VT19060|0|MEMB002B...</td>\n",
       "      <td>53</td>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(VT14350|0|MEMB007D|+|3005,)</td>\n",
       "      <td>50</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>(VT18420|0|MEMB002A|-|2172, VT18420|0|MEMB002B...</td>\n",
       "      <td>49</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(VT18480|1|MEMB004B|-|3211,)</td>\n",
       "      <td>44</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Species with TFBS at Position  Count  Alignment File\n",
       "11                        (VT19240|1|MEMB003F|-|4104,)    109              41\n",
       "0    (VT19020|1|MEMB003A|-|4766, VT19020|1|MEMB003D...     94              37\n",
       "0                         (VT10600|0|MEMB002F|-|2671,)     84               2\n",
       "3                         (VT18830|1|MEMB002F|+|3408,)     82              36\n",
       "32                        (VT14350|0|MEMB002F|+|3647,)     72               6\n",
       "6    (VT19750|1|MEMB002A|+|2196, VT19750|1|MEMB002B...     67              45\n",
       "3    (VT19600|0|MEMB002A|-|2450, VT19600|0|MEMB002B...     63              44\n",
       "2    (VT19350|0|MEMB002A|+|2674, VT19350|0|MEMB002B...     56              43\n",
       "1    (VT17000|1|MEMB002A|-|2389, VT17000|1|MEMB002B...     56              20\n",
       "0                         (VT19920|1|MEMB003B|+|3304,)     56              48\n",
       "1    (VT14350|0|MEMB002B|+|3460, VT14350|0|MEMB004A...     53               6\n",
       "110  (VT19060|0|MEMB002A|+|2590, VT19060|0|MEMB002B...     53              39\n",
       "2                         (VT14350|0|MEMB007D|+|3005,)     50               6\n",
       "16   (VT18420|0|MEMB002A|-|2172, VT18420|0|MEMB002B...     49              30\n",
       "3                         (VT18480|1|MEMB004B|-|3211,)     44              33"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "freq_tables = []\n",
    "for i in range(len(tables)): # len(tables) represents the alignments\n",
    "    t = frequent_patterns(i)\n",
    "    t[\"Alignment File\"] = [i] * t.count()[0]\n",
    "    freq_tables.append(t)\n",
    "\n",
    "frequent_all_alignments = pd.concat(freq_tables).sort_values(by=['Count'], ascending=False)\n",
    "frequent_all_alignments.head(15)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### There is a combination involving 3 species:\n",
    "* 'VT19020|1|MEMB003A|-|4766'\n",
    "* 'VT19020|1|MEMB003D|+|4904'\n",
    "* 'VT19020|1|dkik|-|4763'\n",
    "#### and it reoccurs 94 times in alignment file 37."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('VT19020|1|MEMB003A|-|4766',\n",
       " 'VT19020|1|MEMB003D|+|4904',\n",
       " 'VT19020|1|dkik|-|4763')"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ID_dictionary(37, 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Wrapping Into One Package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def pipeline(align_path, motif_path, motif, alignment_num, percentile=80):\n",
    "    \"\"\"\n",
    "    align_path: a string specifying the path to alignment files you want to use.\n",
    "    motif_path: a string specifying the path to pwm files you want to use.\n",
    "    motif: integer specifying which motif to use (indexes into list of motifs)\n",
    "    alignment_num: an integer specifying the specific alignment you want to analyze.\n",
    "    percentile: cutoff for high frequency of combination\n",
    "    \n",
    "    Returns:\n",
    "    1) Table with unique alignment position, combinations at positions, and frequencies\n",
    "    2) Table of high-frequency combinations, with high frequency defined as >= ? percentile\n",
    "    3) Table containing information about the most frequency combination\n",
    "    \"\"\"\n",
    "    files = glob(align_path)\n",
    "    all_motifs = glob(motif_path)\n",
    "    \n",
    "    filtered_tables = []\n",
    "    for i in files[:10]: # CHANGE NUMBER\n",
    "        table = calculate_one_dfs_TFBS(i, [all_motifs[motif]])\n",
    "        f = filter_95_percentile(table, \"cad_FlyReg.fm\").drop(['seq_len', 'position'], axis=1)\n",
    "        filtered_tables.append(f)\n",
    "\n",
    "    position_species_tables = []\n",
    "    for f in filtered_tables:\n",
    "        positions_with_TFBS = list(f[\"align_position\"])\n",
    "        combos = {} # Dictionary with position index : list of species that have TFBS at this position\n",
    "        for i in positions_with_TFBS:\n",
    "            combos[i] = list(f[f[\"align_position\"] == i][\"species\"])\n",
    "        d = {'Alignment Positions': list(combos.keys()),\n",
    "             'Species with TFBS at Position': list(combos.values())}\n",
    "        positions_species = pd.DataFrame(data=d)\n",
    "        positions_species[\"Species with TFBS at Position\"] = [tuple(i) for i in positions_species[\"Species with TFBS at Position\"]]\n",
    "        position_species_tables.append(positions_species)\n",
    "\n",
    "    group_by_combo_tables = []\n",
    "    for p in position_species_tables:\n",
    "        group_by_combo = p.groupby(\"Species with TFBS at Position\",sort=False).count().reset_index()\n",
    "        group_by_combo.columns = [\"Species with TFBS at Position\", \"Count\"]\n",
    "        group_by_combo_tables.append(group_by_combo)\n",
    "\n",
    "    counts_at_positions_tables = [] # table to return\n",
    "    num_alignments = len(group_by_combo_tables)\n",
    "    for n in range(num_alignments):\n",
    "        counts_at_positions = position_species_tables[n].merge(group_by_combo_tables[n], how=\"left\",sort=False)\n",
    "        counts_at_positions.set_index(\"Alignment Positions\", inplace=True)\n",
    "        counts_at_positions_tables.append(counts_at_positions)\n",
    "    \n",
    "    counts = list(counts_at_positions_tables[alignment_num][\"Count\"])\n",
    "    g_table = group_by_combo_tables[alignment_num]\n",
    "    freq_patterns = g_table[g_table[\"Count\"] >= np.percentile(counts, percentile)] # return\n",
    "\n",
    "    max_count = max(list(counts_at_positions_tables[alignment_num][\"Count\"]))\n",
    "    table = group_by_combo_tables[alignment_num]\n",
    "    most_freq_pattern = table[table[\"Count\"] == max_count] #return\n",
    "\n",
    "    return counts_at_positions_tables[alignment_num], freq_patterns, most_freq_pattern"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "External functions used in pipeline:\n",
    "* calculate_one_dfs_TFBS\n",
    "* filter_95_percentile\n",
    "\n",
    "Deleted from original map motif:\n",
    "* pwm_threshold\n",
    "* calculate_dfs_TFBS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "results = pipeline('../data/alignments/*0.fa','../data/pwm/*.fm',1,2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alignment Positions</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>520</th>\n",
       "      <td>(VT10600|0|MEMB002F|-|2671,)</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2058</th>\n",
       "      <td>(VT10600|0|MEMB002D|-|719,)</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1548</th>\n",
       "      <td>(VT10600|0|MEMB002F|-|2671,)</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1313</th>\n",
       "      <td>(VT10600|0|MEMB004A|-|935,)</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>(VT10600|0|MEMB002A|-|905, VT10600|0|MEMB002B|...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Species with TFBS at Position  Count\n",
       "Alignment Positions                                                          \n",
       "520                                       (VT10600|0|MEMB002F|-|2671,)     84\n",
       "2058                                       (VT10600|0|MEMB002D|-|719,)      7\n",
       "1548                                      (VT10600|0|MEMB002F|-|2671,)     84\n",
       "1313                                       (VT10600|0|MEMB004A|-|935,)      7\n",
       "17                   (VT10600|0|MEMB002A|-|905, VT10600|0|MEMB002B|...      1"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results[0].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(VT10600|0|MEMB002F|-|2671,)</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Species with TFBS at Position  Count\n",
       "0  (VT10600|0|MEMB002F|-|2671,)     84"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(VT10600|0|MEMB002F|-|2671,)</td>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Species with TFBS at Position  Count\n",
       "0  (VT10600|0|MEMB002F|-|2671,)     84"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
