{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Motif Mapping\n",
    "\n",
    "Using Ciera Martinez's map_motif.py in a generalized jupyter notebook function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "import Bio\n",
    "\n",
    "# For computations\n",
    "import string\n",
    "import random\n",
    "import matplotlib.pyplot as plt; plt.rcdefaults()\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "from IPython.display import Image\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from glob import glob\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython.display import display\n",
    "%matplotlib inline\n",
    "plt.style.use('fivethirtyeight')\n",
    "\n",
    "import os, sys\n",
    "from Bio import motifs\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Alphabet import IUPAC, generic_dna, generic_protein\n",
    "\n",
    "# Import definitions & functions from .py file:\n",
    "from MSE import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Processing\n",
    "\n",
    "Reads in the alligned raw DNA sequences and motifs to be pipelined:\n",
    "\n",
    "1. Saves a copy of the sequence without the '$-$' included.\n",
    "\n",
    "2. Casts the undashed version as IUPACUnambiguousDNA().\n",
    "\n",
    "3. Calculates the thresholds to be used in assigning scores for likihood of the motif appearing.\n",
    "\n",
    "4. Takes the indicies that have been scored and maps them to the corresponding raw sequence index, taking into consideration if the strand was positive or negative.\n",
    "\n",
    "5. Returns a df with $\\scriptsize{[\\textbf{position}, \\textbf{score, sequence length}, \\textbf{species}, \\textbf{raw_position index}, \\textbf{strand positioning}, \\textbf{aligned position}]}$.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Reads through and saves the names of all same files(.fa and .fm respectively) found in the same directory.\n",
    "\n",
    "files = glob('../data/alignments/*0.fa')\n",
    "all_motifs = glob('../data/pwm/*.fm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Extraction\n",
    "\n",
    "Taking the dfs that come from the \"calculate_all_TFBS\" function we filter out those with a score above the 95th percentile. With the filtered out values we match the species with their respective DNA sequences and then locate and save the given motif within the DNA sequence. Whether the strand is positive or negative is treated individually and the -'s are included in the sequence indexing. This newly extracted information is saved as columns and appended to the df."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following 3 cells provide a convenient way to access the standard \"cutoff\" point when using the 4 motifs printed below. cutoff_dict is used in the filter_95_percentile function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['../data/pwm/cad_FlyReg.fm',\n",
       " '../data/pwm/eve_new6.fm',\n",
       " '../data/pwm/zelda_.fm',\n",
       " '../data/pwm/bcd_FlyReg.fm']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "cutoff_dict = {}\n",
    "cutoff_dict[\"bcd_FlyReg.fm\"] = stand_cutoff(all_motifs, 0)[0]\n",
    "cutoff_dict[\"cad_FlyReg.fm\"] = stand_cutoff(all_motifs, 1)[0]\n",
    "cutoff_dict[\"eve_new6.fm\"] = stand_cutoff(all_motifs, 2)[0]\n",
    "cutoff_dict[\"zelda_.fm\"] = stand_cutoff(all_motifs, 3)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'bcd_FlyReg.fm': 2.648324489593506,\n",
       " 'cad_FlyReg.fm': 3.44942045211792,\n",
       " 'eve_new6.fm': 6.678131580352783,\n",
       " 'zelda_.fm': 3.3171534538269043}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cutoff_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n",
      "69\n",
      "70\n",
      "71\n",
      "72\n",
      "73\n",
      "74\n",
      "75\n",
      "76\n",
      "77\n",
      "78\n",
      "79\n",
      "80\n",
      "81\n",
      "82\n",
      "83\n",
      "84\n",
      "85\n",
      "86\n",
      "87\n",
      "88\n",
      "89\n",
      "90\n",
      "91\n",
      "92\n",
      "93\n",
      "94\n",
      "95\n",
      "96\n",
      "97\n",
      "98\n",
      "99\n",
      "100\n",
      "101\n",
      "102\n",
      "103\n",
      "104\n",
      "105\n",
      "106\n",
      "107\n",
      "108\n",
      "109\n",
      "110\n",
      "111\n",
      "112\n",
      "113\n",
      "114\n",
      "115\n",
      "116\n",
      "117\n",
      "118\n",
      "119\n",
      "120\n",
      "121\n",
      "122\n",
      "123\n",
      "124\n",
      "125\n",
      "126\n",
      "127\n",
      "128\n",
      "129\n",
      "130\n",
      "131\n",
      "132\n",
      "133\n",
      "134\n",
      "135\n",
      "136\n",
      "137\n",
      "138\n",
      "139\n",
      "140\n",
      "141\n",
      "142\n",
      "143\n",
      "144\n",
      "145\n",
      "146\n",
      "147\n",
      "148\n",
      "149\n",
      "150\n",
      "151\n",
      "152\n",
      "153\n",
      "154\n",
      "155\n",
      "156\n",
      "157\n",
      "158\n",
      "159\n",
      "160\n",
      "161\n",
      "162\n",
      "163\n",
      "164\n",
      "165\n",
      "166\n",
      "167\n",
      "168\n",
      "169\n",
      "170\n",
      "171\n",
      "172\n",
      "173\n",
      "174\n",
      "175\n",
      "176\n",
      "177\n",
      "178\n",
      "179\n",
      "180\n",
      "181\n",
      "182\n",
      "183\n",
      "184\n",
      "185\n",
      "186\n",
      "187\n",
      "188\n",
      "189\n",
      "190\n",
      "191\n",
      "192\n",
      "193\n",
      "194\n",
      "195\n",
      "196\n",
      "197\n",
      "198\n",
      "199\n",
      "200\n",
      "201\n",
      "202\n",
      "203\n",
      "204\n",
      "205\n",
      "206\n",
      "207\n",
      "208\n",
      "209\n",
      "210\n",
      "211\n",
      "212\n",
      "213\n",
      "214\n",
      "215\n",
      "216\n",
      "217\n",
      "218\n",
      "219\n",
      "220\n",
      "221\n",
      "222\n",
      "223\n",
      "224\n",
      "225\n",
      "226\n",
      "227\n",
      "228\n",
      "229\n",
      "230\n",
      "231\n",
      "232\n",
      "233\n",
      "234\n",
      "235\n",
      "236\n",
      "237\n",
      "238\n",
      "239\n",
      "240\n",
      "241\n",
      "242\n",
      "243\n",
      "244\n",
      "245\n",
      "246\n",
      "247\n",
      "248\n",
      "249\n",
      "250\n",
      "251\n",
      "252\n",
      "253\n",
      "254\n",
      "255\n",
      "256\n",
      "257\n",
      "258\n",
      "259\n",
      "260\n",
      "261\n",
      "262\n",
      "263\n",
      "264\n",
      "265\n",
      "266\n",
      "267\n",
      "268\n",
      "269\n",
      "270\n",
      "271\n",
      "272\n",
      "273\n",
      "274\n",
      "275\n",
      "276\n",
      "277\n",
      "278\n",
      "279\n",
      "280\n",
      "281\n",
      "282\n",
      "283\n",
      "284\n",
      "285\n",
      "286\n",
      "287\n",
      "288\n",
      "289\n",
      "290\n",
      "291\n",
      "292\n",
      "293\n",
      "294\n",
      "295\n",
      "296\n",
      "297\n",
      "298\n",
      "299\n",
      "300\n",
      "301\n",
      "302\n",
      "303\n",
      "304\n",
      "305\n",
      "306\n",
      "307\n",
      "308\n",
      "309\n",
      "310\n",
      "311\n",
      "312\n",
      "313\n",
      "314\n",
      "315\n",
      "316\n",
      "317\n",
      "318\n",
      "319\n",
      "320\n",
      "321\n",
      "322\n",
      "323\n",
      "324\n",
      "325\n",
      "326\n",
      "327\n",
      "328\n",
      "329\n",
      "330\n",
      "331\n",
      "332\n",
      "333\n",
      "334\n",
      "335\n",
      "336\n",
      "337\n",
      "338\n",
      "339\n",
      "340\n",
      "341\n",
      "342\n",
      "343\n",
      "344\n",
      "345\n",
      "346\n",
      "347\n",
      "348\n",
      "349\n",
      "350\n",
      "351\n",
      "352\n",
      "353\n",
      "354\n",
      "355\n",
      "356\n",
      "357\n",
      "358\n",
      "359\n",
      "360\n",
      "361\n"
     ]
    }
   ],
   "source": [
    "# LOOP THROUGH & APPLY TO ALL THE ALIGNMENTS FOR Zelda\n",
    "## Count is just to \n",
    "count=0\n",
    "tables = []\n",
    "for i in files[:500]:\n",
    "    count = count + 1\n",
    "    tables.append(calculate_one_dfs_TFBS(i, [all_motifs[2]]))\n",
    "    print(count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# LOOP THROUGH & APPLY TO ALL THE ALIGNMENTS\n",
    "filtered_tables = []\n",
    "for i in tables:\n",
    "    f = filter_95_percentile(i, \"zelda_.fm\").drop(['seq_len', 'position'], axis=1)\n",
    "    filtered_tables.append(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Grouping Sequences by species that share TFBS presence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtered_tables\n",
    "* filtered_tables is a LIST of TABLES; each table in this list represents ONE alignment file. In every table, 1 row represents a spot on the sequence that we've chosen to label as TFBS, based on our 95% criteria. You can index into filtered_tables to view any one table. For example, alignment 1:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### position_species_tables\n",
    "(See below for example) position_species_tables is also a LIST of TABLES. Again, each element in the list represents ONE alignment file. Each table tells you:\n",
    "\n",
    "1. The alignment position\n",
    "2. A tuple containing all the species that have TFBS's at that position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>species</th>\n",
       "      <th>raw_position</th>\n",
       "      <th>strand</th>\n",
       "      <th>align_position</th>\n",
       "      <th>motif</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>603</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>33</td>\n",
       "      <td>positive</td>\n",
       "      <td>65</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>605</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>72</td>\n",
       "      <td>positive</td>\n",
       "      <td>138</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>608</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>134</td>\n",
       "      <td>negative</td>\n",
       "      <td>221</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>609</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>150</td>\n",
       "      <td>positive</td>\n",
       "      <td>237</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>184</td>\n",
       "      <td>positive</td>\n",
       "      <td>271</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>615</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>217</td>\n",
       "      <td>negative</td>\n",
       "      <td>326</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>619</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>350</td>\n",
       "      <td>negative</td>\n",
       "      <td>527</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>626</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>577</td>\n",
       "      <td>negative</td>\n",
       "      <td>782</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>634</td>\n",
       "      <td>negative</td>\n",
       "      <td>841</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>630</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>704</td>\n",
       "      <td>positive</td>\n",
       "      <td>919</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>638</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>928</td>\n",
       "      <td>positive</td>\n",
       "      <td>1196</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>945</td>\n",
       "      <td>negative</td>\n",
       "      <td>1220</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>640</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1009</td>\n",
       "      <td>positive</td>\n",
       "      <td>1302</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>641</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1030</td>\n",
       "      <td>negative</td>\n",
       "      <td>1462</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>642</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1036</td>\n",
       "      <td>negative</td>\n",
       "      <td>1499</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>643</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1050</td>\n",
       "      <td>positive</td>\n",
       "      <td>1514</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1071</td>\n",
       "      <td>negative</td>\n",
       "      <td>1551</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1154</td>\n",
       "      <td>positive</td>\n",
       "      <td>1670</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1161</td>\n",
       "      <td>positive</td>\n",
       "      <td>1677</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>650</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1208</td>\n",
       "      <td>negative</td>\n",
       "      <td>1727</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1237</td>\n",
       "      <td>positive</td>\n",
       "      <td>1763</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>653</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1251</td>\n",
       "      <td>positive</td>\n",
       "      <td>1777</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>656</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1302</td>\n",
       "      <td>negative</td>\n",
       "      <td>1841</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1309</td>\n",
       "      <td>positive</td>\n",
       "      <td>1848</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>658</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1327</td>\n",
       "      <td>positive</td>\n",
       "      <td>1868</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>664</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1462</td>\n",
       "      <td>positive</td>\n",
       "      <td>2086</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>671</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1611</td>\n",
       "      <td>negative</td>\n",
       "      <td>2273</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>672</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1617</td>\n",
       "      <td>negative</td>\n",
       "      <td>2280</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>676</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1828</td>\n",
       "      <td>negative</td>\n",
       "      <td>2607</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>677</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|MEMB002A|+|2656</td>\n",
       "      <td>1836</td>\n",
       "      <td>negative</td>\n",
       "      <td>2615</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1928</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1183</td>\n",
       "      <td>positive</td>\n",
       "      <td>1848</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1929</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1201</td>\n",
       "      <td>positive</td>\n",
       "      <td>1868</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1935</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1342</td>\n",
       "      <td>positive</td>\n",
       "      <td>2086</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1937</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1379</td>\n",
       "      <td>positive</td>\n",
       "      <td>2125</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1940</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1469</td>\n",
       "      <td>negative</td>\n",
       "      <td>2243</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1942</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1499</td>\n",
       "      <td>negative</td>\n",
       "      <td>2273</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1943</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1505</td>\n",
       "      <td>negative</td>\n",
       "      <td>2280</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1944</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1612</td>\n",
       "      <td>negative</td>\n",
       "      <td>2412</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1948</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1783</td>\n",
       "      <td>negative</td>\n",
       "      <td>2740</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1951</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1813</td>\n",
       "      <td>positive</td>\n",
       "      <td>2777</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1952</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1839</td>\n",
       "      <td>positive</td>\n",
       "      <td>2809</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1953</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1854</td>\n",
       "      <td>negative</td>\n",
       "      <td>2827</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1956</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1865</td>\n",
       "      <td>positive</td>\n",
       "      <td>2840</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1959</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1971</td>\n",
       "      <td>positive</td>\n",
       "      <td>3026</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1961</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1978</td>\n",
       "      <td>positive</td>\n",
       "      <td>3033</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1962</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>1993</td>\n",
       "      <td>negative</td>\n",
       "      <td>3053</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1966</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>2101</td>\n",
       "      <td>negative</td>\n",
       "      <td>3232</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1967</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>2123</td>\n",
       "      <td>positive</td>\n",
       "      <td>3254</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1969</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>2139</td>\n",
       "      <td>negative</td>\n",
       "      <td>3272</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1977</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>2387</td>\n",
       "      <td>negative</td>\n",
       "      <td>3571</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1978</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>2393</td>\n",
       "      <td>negative</td>\n",
       "      <td>3577</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1986</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>2519</td>\n",
       "      <td>positive</td>\n",
       "      <td>3712</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1987</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>2528</td>\n",
       "      <td>negative</td>\n",
       "      <td>3730</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1990</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>2562</td>\n",
       "      <td>negative</td>\n",
       "      <td>3791</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1991</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>2638</td>\n",
       "      <td>positive</td>\n",
       "      <td>4029</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1992</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>2665</td>\n",
       "      <td>negative</td>\n",
       "      <td>4056</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1993</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>2671</td>\n",
       "      <td>negative</td>\n",
       "      <td>4062</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>2847</td>\n",
       "      <td>positive</td>\n",
       "      <td>4474</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2005</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>2924</td>\n",
       "      <td>positive</td>\n",
       "      <td>4555</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2006</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>VT8450|1|dkik|+|2972</td>\n",
       "      <td>2930</td>\n",
       "      <td>positive</td>\n",
       "      <td>4561</td>\n",
       "      <td>zelda_.fm</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>897 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         score                   species  raw_position    strand  \\\n",
       "603   6.678132  VT8450|1|MEMB002A|+|2656            33  positive   \n",
       "605   6.678132  VT8450|1|MEMB002A|+|2656            72  positive   \n",
       "608   6.678132  VT8450|1|MEMB002A|+|2656           134  negative   \n",
       "609   6.678132  VT8450|1|MEMB002A|+|2656           150  positive   \n",
       "614   6.678132  VT8450|1|MEMB002A|+|2656           184  positive   \n",
       "615   6.678132  VT8450|1|MEMB002A|+|2656           217  negative   \n",
       "619   6.678132  VT8450|1|MEMB002A|+|2656           350  negative   \n",
       "626   6.678132  VT8450|1|MEMB002A|+|2656           577  negative   \n",
       "628   6.678132  VT8450|1|MEMB002A|+|2656           634  negative   \n",
       "630   6.678132  VT8450|1|MEMB002A|+|2656           704  positive   \n",
       "638   6.678132  VT8450|1|MEMB002A|+|2656           928  positive   \n",
       "639   6.678132  VT8450|1|MEMB002A|+|2656           945  negative   \n",
       "640   6.678132  VT8450|1|MEMB002A|+|2656          1009  positive   \n",
       "641   6.678132  VT8450|1|MEMB002A|+|2656          1030  negative   \n",
       "642   6.678132  VT8450|1|MEMB002A|+|2656          1036  negative   \n",
       "643   6.678132  VT8450|1|MEMB002A|+|2656          1050  positive   \n",
       "644   6.678132  VT8450|1|MEMB002A|+|2656          1071  negative   \n",
       "645   6.678132  VT8450|1|MEMB002A|+|2656          1154  positive   \n",
       "646   6.678132  VT8450|1|MEMB002A|+|2656          1161  positive   \n",
       "650   6.678132  VT8450|1|MEMB002A|+|2656          1208  negative   \n",
       "651   6.678132  VT8450|1|MEMB002A|+|2656          1237  positive   \n",
       "653   6.678132  VT8450|1|MEMB002A|+|2656          1251  positive   \n",
       "656   6.678132  VT8450|1|MEMB002A|+|2656          1302  negative   \n",
       "657   6.678132  VT8450|1|MEMB002A|+|2656          1309  positive   \n",
       "658   6.678132  VT8450|1|MEMB002A|+|2656          1327  positive   \n",
       "664   6.678132  VT8450|1|MEMB002A|+|2656          1462  positive   \n",
       "671   6.678132  VT8450|1|MEMB002A|+|2656          1611  negative   \n",
       "672   6.678132  VT8450|1|MEMB002A|+|2656          1617  negative   \n",
       "676   6.678132  VT8450|1|MEMB002A|+|2656          1828  negative   \n",
       "677   6.678132  VT8450|1|MEMB002A|+|2656          1836  negative   \n",
       "...        ...                       ...           ...       ...   \n",
       "1928  6.678132      VT8450|1|dkik|+|2972          1183  positive   \n",
       "1929  6.678132      VT8450|1|dkik|+|2972          1201  positive   \n",
       "1935  6.678132      VT8450|1|dkik|+|2972          1342  positive   \n",
       "1937  6.678132      VT8450|1|dkik|+|2972          1379  positive   \n",
       "1940  6.678132      VT8450|1|dkik|+|2972          1469  negative   \n",
       "1942  6.678132      VT8450|1|dkik|+|2972          1499  negative   \n",
       "1943  6.678132      VT8450|1|dkik|+|2972          1505  negative   \n",
       "1944  6.678132      VT8450|1|dkik|+|2972          1612  negative   \n",
       "1948  6.678132      VT8450|1|dkik|+|2972          1783  negative   \n",
       "1951  6.678132      VT8450|1|dkik|+|2972          1813  positive   \n",
       "1952  6.678132      VT8450|1|dkik|+|2972          1839  positive   \n",
       "1953  6.678132      VT8450|1|dkik|+|2972          1854  negative   \n",
       "1956  6.678132      VT8450|1|dkik|+|2972          1865  positive   \n",
       "1959  6.678132      VT8450|1|dkik|+|2972          1971  positive   \n",
       "1961  6.678132      VT8450|1|dkik|+|2972          1978  positive   \n",
       "1962  6.678132      VT8450|1|dkik|+|2972          1993  negative   \n",
       "1966  6.678132      VT8450|1|dkik|+|2972          2101  negative   \n",
       "1967  6.678132      VT8450|1|dkik|+|2972          2123  positive   \n",
       "1969  6.678132      VT8450|1|dkik|+|2972          2139  negative   \n",
       "1977  6.678132      VT8450|1|dkik|+|2972          2387  negative   \n",
       "1978  6.678132      VT8450|1|dkik|+|2972          2393  negative   \n",
       "1986  6.678132      VT8450|1|dkik|+|2972          2519  positive   \n",
       "1987  6.678132      VT8450|1|dkik|+|2972          2528  negative   \n",
       "1990  6.678132      VT8450|1|dkik|+|2972          2562  negative   \n",
       "1991  6.678132      VT8450|1|dkik|+|2972          2638  positive   \n",
       "1992  6.678132      VT8450|1|dkik|+|2972          2665  negative   \n",
       "1993  6.678132      VT8450|1|dkik|+|2972          2671  negative   \n",
       "1998  6.678132      VT8450|1|dkik|+|2972          2847  positive   \n",
       "2005  6.678132      VT8450|1|dkik|+|2972          2924  positive   \n",
       "2006  6.678132      VT8450|1|dkik|+|2972          2930  positive   \n",
       "\n",
       "      align_position      motif  \n",
       "603               65  zelda_.fm  \n",
       "605              138  zelda_.fm  \n",
       "608              221  zelda_.fm  \n",
       "609              237  zelda_.fm  \n",
       "614              271  zelda_.fm  \n",
       "615              326  zelda_.fm  \n",
       "619              527  zelda_.fm  \n",
       "626              782  zelda_.fm  \n",
       "628              841  zelda_.fm  \n",
       "630              919  zelda_.fm  \n",
       "638             1196  zelda_.fm  \n",
       "639             1220  zelda_.fm  \n",
       "640             1302  zelda_.fm  \n",
       "641             1462  zelda_.fm  \n",
       "642             1499  zelda_.fm  \n",
       "643             1514  zelda_.fm  \n",
       "644             1551  zelda_.fm  \n",
       "645             1670  zelda_.fm  \n",
       "646             1677  zelda_.fm  \n",
       "650             1727  zelda_.fm  \n",
       "651             1763  zelda_.fm  \n",
       "653             1777  zelda_.fm  \n",
       "656             1841  zelda_.fm  \n",
       "657             1848  zelda_.fm  \n",
       "658             1868  zelda_.fm  \n",
       "664             2086  zelda_.fm  \n",
       "671             2273  zelda_.fm  \n",
       "672             2280  zelda_.fm  \n",
       "676             2607  zelda_.fm  \n",
       "677             2615  zelda_.fm  \n",
       "...              ...        ...  \n",
       "1928            1848  zelda_.fm  \n",
       "1929            1868  zelda_.fm  \n",
       "1935            2086  zelda_.fm  \n",
       "1937            2125  zelda_.fm  \n",
       "1940            2243  zelda_.fm  \n",
       "1942            2273  zelda_.fm  \n",
       "1943            2280  zelda_.fm  \n",
       "1944            2412  zelda_.fm  \n",
       "1948            2740  zelda_.fm  \n",
       "1951            2777  zelda_.fm  \n",
       "1952            2809  zelda_.fm  \n",
       "1953            2827  zelda_.fm  \n",
       "1956            2840  zelda_.fm  \n",
       "1959            3026  zelda_.fm  \n",
       "1961            3033  zelda_.fm  \n",
       "1962            3053  zelda_.fm  \n",
       "1966            3232  zelda_.fm  \n",
       "1967            3254  zelda_.fm  \n",
       "1969            3272  zelda_.fm  \n",
       "1977            3571  zelda_.fm  \n",
       "1978            3577  zelda_.fm  \n",
       "1986            3712  zelda_.fm  \n",
       "1987            3730  zelda_.fm  \n",
       "1990            3791  zelda_.fm  \n",
       "1991            4029  zelda_.fm  \n",
       "1992            4056  zelda_.fm  \n",
       "1993            4062  zelda_.fm  \n",
       "1998            4474  zelda_.fm  \n",
       "2005            4555  zelda_.fm  \n",
       "2006            4561  zelda_.fm  \n",
       "\n",
       "[897 rows x 6 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_tables[6]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Splitting the species column\n",
    "\n",
    "In order to simplify the species grouping we are going to isolate the species name. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Loop through all the filtered dataframes to extract informative information\n",
    "## from sequence ID, makes things cleaner\n",
    "## ONLY RUN ONCE, IF YOU NEED TO RE_RUN MAKE NEW filtered_tables\n",
    "\n",
    "for df in filtered_tables:\n",
    "    df['align_file'] = df['species'].str.split('|').str[0]\n",
    "    df['function'] = df['species'].str.split('|').str[1]\n",
    "    df['species'] = df['species'].str.split('|').str[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>species</th>\n",
       "      <th>raw_position</th>\n",
       "      <th>strand</th>\n",
       "      <th>align_position</th>\n",
       "      <th>motif</th>\n",
       "      <th>align_file</th>\n",
       "      <th>function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1748</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>MEMB002A</td>\n",
       "      <td>24</td>\n",
       "      <td>negative</td>\n",
       "      <td>45</td>\n",
       "      <td>zelda_.fm</td>\n",
       "      <td>VT29210</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1765</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>MEMB002A</td>\n",
       "      <td>337</td>\n",
       "      <td>negative</td>\n",
       "      <td>1218</td>\n",
       "      <td>zelda_.fm</td>\n",
       "      <td>VT29210</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1767</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>MEMB002A</td>\n",
       "      <td>392</td>\n",
       "      <td>negative</td>\n",
       "      <td>1400</td>\n",
       "      <td>zelda_.fm</td>\n",
       "      <td>VT29210</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1772</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>MEMB002A</td>\n",
       "      <td>456</td>\n",
       "      <td>negative</td>\n",
       "      <td>1515</td>\n",
       "      <td>zelda_.fm</td>\n",
       "      <td>VT29210</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1775</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>MEMB002A</td>\n",
       "      <td>495</td>\n",
       "      <td>negative</td>\n",
       "      <td>1568</td>\n",
       "      <td>zelda_.fm</td>\n",
       "      <td>VT29210</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1776</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>MEMB002A</td>\n",
       "      <td>504</td>\n",
       "      <td>negative</td>\n",
       "      <td>1577</td>\n",
       "      <td>zelda_.fm</td>\n",
       "      <td>VT29210</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1779</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>MEMB002A</td>\n",
       "      <td>514</td>\n",
       "      <td>positive</td>\n",
       "      <td>1598</td>\n",
       "      <td>zelda_.fm</td>\n",
       "      <td>VT29210</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1783</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>MEMB002A</td>\n",
       "      <td>552</td>\n",
       "      <td>negative</td>\n",
       "      <td>1637</td>\n",
       "      <td>zelda_.fm</td>\n",
       "      <td>VT29210</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1787</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>MEMB002A</td>\n",
       "      <td>627</td>\n",
       "      <td>positive</td>\n",
       "      <td>1719</td>\n",
       "      <td>zelda_.fm</td>\n",
       "      <td>VT29210</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1789</th>\n",
       "      <td>6.678132</td>\n",
       "      <td>MEMB002A</td>\n",
       "      <td>640</td>\n",
       "      <td>positive</td>\n",
       "      <td>1732</td>\n",
       "      <td>zelda_.fm</td>\n",
       "      <td>VT29210</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         score   species  raw_position    strand  align_position      motif  \\\n",
       "1748  6.678132  MEMB002A            24  negative              45  zelda_.fm   \n",
       "1765  6.678132  MEMB002A           337  negative            1218  zelda_.fm   \n",
       "1767  6.678132  MEMB002A           392  negative            1400  zelda_.fm   \n",
       "1772  6.678132  MEMB002A           456  negative            1515  zelda_.fm   \n",
       "1775  6.678132  MEMB002A           495  negative            1568  zelda_.fm   \n",
       "1776  6.678132  MEMB002A           504  negative            1577  zelda_.fm   \n",
       "1779  6.678132  MEMB002A           514  positive            1598  zelda_.fm   \n",
       "1783  6.678132  MEMB002A           552  negative            1637  zelda_.fm   \n",
       "1787  6.678132  MEMB002A           627  positive            1719  zelda_.fm   \n",
       "1789  6.678132  MEMB002A           640  positive            1732  zelda_.fm   \n",
       "\n",
       "     align_file function  \n",
       "1748    VT29210        0  \n",
       "1765    VT29210        0  \n",
       "1767    VT29210        0  \n",
       "1772    VT29210        0  \n",
       "1775    VT29210        0  \n",
       "1776    VT29210        0  \n",
       "1779    VT29210        0  \n",
       "1783    VT29210        0  \n",
       "1787    VT29210        0  \n",
       "1789    VT29210        0  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_tables[3].head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "## Write out filtered_tables \n",
    "pd.concat(filtered_tables).to_csv('~/Desktop/filtered_tables_zelda_.fm.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "position_species_tables = []\n",
    "\n",
    "for f in filtered_tables:\n",
    "    positions_with_TFBS = list(f[\"align_position\"])\n",
    "    combos = {} # Dictionary with position index : list of species that have TFBS at this position\n",
    "    for i in positions_with_TFBS:\n",
    "        combos[i] = list(f[f[\"align_position\"] == i][\"species\"])\n",
    "\n",
    "    d = {'Alignment Positions': list(combos.keys()),\n",
    "         'Species with TFBS at Position': list(combos.values())}\n",
    "    positions_species = pd.DataFrame(data=d)\n",
    "    positions_species[\"Species with TFBS at Position\"] = [tuple(i) for i in positions_species[\"Species with TFBS at Position\"]]\n",
    "    # positions_species[\"align_file\"] = filtered_tables[0].iloc[0][6] # add align_file_ID\n",
    "    # positions_species[\"function\"] = filtered_tables[0].iloc[0][7] # add function\n",
    "    \n",
    "    position_species_tables.append(positions_species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Alignment Positions</th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>230</td>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>303</td>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>321</td>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>329</td>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB003C, MEMB0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>331</td>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>350</td>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>388</td>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB003C, MEMB0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>395</td>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>797</td>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB003A, MEMB003D, MEMB0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>811</td>\n",
       "      <td>(MEMB002A, MEMB002C, MEMB003C, MEMB007B)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Alignment Positions                      Species with TFBS at Position\n",
       "0                  230  (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...\n",
       "1                  303  (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...\n",
       "2                  321  (MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...\n",
       "3                  329  (MEMB002A, MEMB002B, MEMB002C, MEMB003C, MEMB0...\n",
       "4                  331  (MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...\n",
       "5                  350  (MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...\n",
       "6                  388  (MEMB002A, MEMB002B, MEMB002C, MEMB003C, MEMB0...\n",
       "7                  395  (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...\n",
       "8                  797  (MEMB002A, MEMB002B, MEMB003A, MEMB003D, MEMB0...\n",
       "9                  811           (MEMB002A, MEMB002C, MEMB003C, MEMB007B)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "position_species_tables[0].head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### group_by_combo_tables\n",
    "Another LIST of tables. Similar to position_species_tables, except each table provides:\n",
    "\n",
    "1. Index: the ID of each unique combination (starts from 0, 1, 2.....)\n",
    "2. Species with TFBS at Position: a tuple containing a list of species, in other words, 1 combination.\n",
    "3. Count: The frequency at which each combination occurs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "group_by_combo_tables = []\n",
    "\n",
    "for p in position_species_tables:\n",
    "    group_by_combo = p.groupby(\"Species with TFBS at Position\",sort=False).count().reset_index()\n",
    "    group_by_combo.columns = [\"Species with TFBS at Position\", \"Count\"]\n",
    "    group_by_combo_tables.append(group_by_combo)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB003C, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB003C, MEMB0...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB003A, MEMB003D, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>(MEMB002A, MEMB002C, MEMB003C, MEMB007B)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Species with TFBS at Position  Count\n",
       "0  (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...      8\n",
       "1  (MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...      1\n",
       "2  (MEMB002A, MEMB002B, MEMB002C, MEMB003C, MEMB0...      1\n",
       "3  (MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...      1\n",
       "4  (MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...      1\n",
       "5  (MEMB002A, MEMB002B, MEMB002C, MEMB003C, MEMB0...      2\n",
       "6  (MEMB002A, MEMB002B, MEMB003A, MEMB003D, MEMB0...      1\n",
       "7           (MEMB002A, MEMB002C, MEMB003C, MEMB007B)      4\n",
       "8  (MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...      1\n",
       "9  (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...      1"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_by_combo_tables[0].head(10)\n",
    "## group_by_combo_tables[0].shape # There are 389 unique IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pd.concat(group_by_combo_tables).to_csv('~/Desktop/group_by_combo_tables_zelda.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "counts_at_positions_tables = []\n",
    "num_alignments = len(group_by_combo_tables)\n",
    "\n",
    "for n in range(num_alignments):\n",
    "    counts_at_positions = position_species_tables[n].merge(group_by_combo_tables[n], how=\"left\",sort=False)\n",
    "    counts_at_positions.set_index(\"Alignment Positions\", inplace=True)\n",
    "    counts_at_positions_tables.append(counts_at_positions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alignment Positions</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB003C, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         Species with TFBS at Position  Count\n",
       "Alignment Positions                                                          \n",
       "230                  (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...      8\n",
       "303                  (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...      8\n",
       "321                  (MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...      1\n",
       "329                  (MEMB002A, MEMB002B, MEMB002C, MEMB003C, MEMB0...      1\n",
       "331                  (MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...      1"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_at_positions_tables[0].head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "position_to_count_dictionaries = []\n",
    "\n",
    "for c in counts_at_positions_tables:\n",
    "    position = list(c.index.values)\n",
    "    dictionary = {}\n",
    "    for i in range(len(position)):\n",
    "        if i in position:\n",
    "            dictionary[i] = c[\"Count\"][i]\n",
    "        else:\n",
    "            dictionary[i] = 0\n",
    "    position_to_count_dictionaries.append(dictionary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def freq_at_position(align_position, alignment):\n",
    "    \"\"\"\n",
    "    align_position: integer\n",
    "    alignment: integer\n",
    "    \n",
    "    Returns the frequency of the pattern occurring at the specified alignment position.\n",
    "    Must specify which alignment file is being used.\n",
    "    \"\"\"\n",
    "    return position_to_count_dictionaries[alignment][align_position]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example: this is the frequency of the pattern occuring at position 12 in the first alignment.\n",
    "freq_at_position(12, 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x11a1bf128>"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos = list(position_to_count_dictionaries[0].keys())\n",
    "y_pos = np.arange(len(pos))\n",
    "freqs = list(position_to_count_dictionaries[0].values())\n",
    " \n",
    "plt.bar(y_pos, freqs, align='center', alpha=0.5, width = 3)\n",
    "plt.xlabel(\"Alignment Position in Sequence Alignment #1\")\n",
    "plt.ylabel(\"Frequency of TFBS Combination at Position\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# For convenience, here are dictionaries of IDs that map to a combination.\n",
    "\n",
    "all_combo_ids = []\n",
    "\n",
    "for g in group_by_combo_tables:\n",
    "    combo_ids = {}\n",
    "    for i in g.index.values:\n",
    "        combo = g[\"Species with TFBS at Position\"][i]\n",
    "        combo_ids[i] = combo\n",
    "    all_combo_ids.append(combo_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def ID_dictionary(alignment, ID):\n",
    "    \"\"\"\n",
    "    alignment: integer specifying which alignment to analyze\n",
    "    ID: integer; ID linked to the unique TFBS combination\n",
    "    \"\"\"\n",
    "    return all_combo_ids[alignment][ID]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('MEMB002A', 'MEMB002B', 'MEMB002C', 'MEMB002D', 'MEMB004A', 'MEMB007B')"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Example: from alignment 2, ID #11 represents this combination of species.\n",
    "ID_dictionary(2, 11)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(MEMB002A, MEMB002C, MEMB003B, MEMB003C, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(MEMB002A, MEMB002C, MEMB002E, MEMB003F)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(MEMB002A, MEMB002C, MEMB003C, MEMB005D, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>(MEMB002A, MEMB002C, MEMB002D, MEMB002E, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>(MEMB002A, MEMB002C, MEMB002F, MEMB003C, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>(MEMB002A, MEMB002C, MEMB003C)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>(MEMB002A, MEMB002C, MEMB002D, MEMB002E, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB003A, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>(MEMB002A, MEMB004A, MEMB006A, MEMB008C)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>(MEMB002A, MEMB002C, MEMB003C, MEMB007B)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>(MEMB002A, MEMB002C, MEMB002F, MEMB003C, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB003C, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>(MEMB002A, MEMB002C, MEMB002F, MEMB003A, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>(MEMB002A, MEMB002C, MEMB002D, MEMB002E, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002E, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>(MEMB002A, MEMB007C)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>(MEMB002A, MEMB003C, MEMB007D)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>(MEMB002B, MEMB004A)</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>(MEMB002B, MEMB002F, MEMB004A, MEMB005B, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>(MEMB002B, MEMB002F, MEMB003B, MEMB003D, MEMB0...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>(MEMB003B, MEMB007C)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>(MEMB003B, MEMB003D, MEMB006A, MEMB007C)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>(MEMB003C,)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>(MEMB003D, MEMB007C)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>(MEMB003D, dkik)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>(MEMB003D,)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>(MEMB003F,)</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>(MEMB003F, MEMB008C)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>(MEMB003F, MEMB006A)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>(MEMB003F, MEMB005D)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>(MEMB004B,)</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>(MEMB004B, MEMB005B, MEMB008C)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>(MEMB004B, MEMB005B)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>(MEMB004E,)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>(MEMB004E, MEMB005D)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>(MEMB004E, MEMB006B)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>(MEMB005B,)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>(MEMB005B, dkik)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>(MEMB005D, MEMB006A)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>(MEMB005D,)</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>(MEMB006A,)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>(MEMB006B,)</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>(MEMB006B, MEMB008C)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>(MEMB006B, MEMB006C)</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>(MEMB006C,)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>(MEMB007B,)</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>(MEMB007C,)</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>(MEMB007D,)</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>(MEMB008C,)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>(dkik,)</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Species with TFBS at Position  Count\n",
       "0    (MEMB002A, MEMB002C, MEMB003B, MEMB003C, MEMB0...      1\n",
       "1    (MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...      1\n",
       "2    (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...      1\n",
       "3             (MEMB002A, MEMB002C, MEMB002E, MEMB003F)      1\n",
       "4    (MEMB002A, MEMB002C, MEMB003C, MEMB005D, MEMB0...      1\n",
       "5    (MEMB002A, MEMB002C, MEMB002D, MEMB002E, MEMB0...      1\n",
       "6    (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...      1\n",
       "7    (MEMB002A, MEMB002C, MEMB002F, MEMB003C, MEMB0...      1\n",
       "8                       (MEMB002A, MEMB002C, MEMB003C)      2\n",
       "9    (MEMB002A, MEMB002C, MEMB002D, MEMB002E, MEMB0...      1\n",
       "10   (MEMB002A, MEMB002B, MEMB002C, MEMB003A, MEMB0...      1\n",
       "11            (MEMB002A, MEMB004A, MEMB006A, MEMB008C)      1\n",
       "12   (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...      1\n",
       "13            (MEMB002A, MEMB002C, MEMB003C, MEMB007B)      2\n",
       "14   (MEMB002A, MEMB002C, MEMB002F, MEMB003C, MEMB0...      1\n",
       "15   (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...      1\n",
       "16   (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...      1\n",
       "17   (MEMB002A, MEMB002B, MEMB002C, MEMB002F, MEMB0...      1\n",
       "18   (MEMB002A, MEMB002B, MEMB002C, MEMB003C, MEMB0...      1\n",
       "19   (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...      1\n",
       "20   (MEMB002A, MEMB002C, MEMB002F, MEMB003A, MEMB0...      1\n",
       "21   (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...      1\n",
       "22   (MEMB002A, MEMB002C, MEMB002D, MEMB002E, MEMB0...      1\n",
       "23   (MEMB002A, MEMB002B, MEMB002C, MEMB002E, MEMB0...      1\n",
       "24   (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...      1\n",
       "25                                (MEMB002A, MEMB007C)      1\n",
       "26                      (MEMB002A, MEMB003C, MEMB007D)      1\n",
       "27                                (MEMB002B, MEMB004A)      3\n",
       "28   (MEMB002B, MEMB002F, MEMB004A, MEMB005B, MEMB0...      1\n",
       "29   (MEMB002B, MEMB002F, MEMB003B, MEMB003D, MEMB0...      1\n",
       "..                                                 ...    ...\n",
       "74                                (MEMB003B, MEMB007C)      1\n",
       "75            (MEMB003B, MEMB003D, MEMB006A, MEMB007C)      1\n",
       "76                                         (MEMB003C,)      1\n",
       "77                                (MEMB003D, MEMB007C)      1\n",
       "78                                    (MEMB003D, dkik)      2\n",
       "79                                         (MEMB003D,)      4\n",
       "80                                         (MEMB003F,)      3\n",
       "81                                (MEMB003F, MEMB008C)      1\n",
       "82                                (MEMB003F, MEMB006A)      1\n",
       "83                                (MEMB003F, MEMB005D)      1\n",
       "84                                         (MEMB004B,)      3\n",
       "85                      (MEMB004B, MEMB005B, MEMB008C)      1\n",
       "86                                (MEMB004B, MEMB005B)      1\n",
       "87                                         (MEMB004E,)      4\n",
       "88                                (MEMB004E, MEMB005D)      1\n",
       "89                                (MEMB004E, MEMB006B)      1\n",
       "90                                         (MEMB005B,)      2\n",
       "91                                    (MEMB005B, dkik)      1\n",
       "92                                (MEMB005D, MEMB006A)      1\n",
       "93                                         (MEMB005D,)      3\n",
       "94                                         (MEMB006A,)      4\n",
       "95                                         (MEMB006B,)      4\n",
       "96                                (MEMB006B, MEMB008C)      1\n",
       "97                                (MEMB006B, MEMB006C)      1\n",
       "98                                         (MEMB006C,)      2\n",
       "99                                         (MEMB007B,)      6\n",
       "100                                        (MEMB007C,)      9\n",
       "101                                        (MEMB007D,)     44\n",
       "102                                        (MEMB008C,)      2\n",
       "103                                            (dkik,)      2\n",
       "\n",
       "[104 rows x 2 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def frequent_patterns(alignment_num, percentile=0):\n",
    "    \"\"\"\n",
    "    alignment_num: integer specifying which alignment to analyze\n",
    "    percentile: (integer) limit frequencies to only above this percentile\n",
    "    \n",
    "    Returns:\n",
    "    A table, like those in group_by_combo_tables, except w/only the largest counts\n",
    "    \"\"\"\n",
    "    counts = list(counts_at_positions_tables[alignment_num][\"Count\"])\n",
    "    table = group_by_combo_tables[alignment_num]\n",
    "    return table[table[\"Count\"] >= np.percentile(counts, percentile)]\n",
    "\n",
    "frequent_patterns(7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "      <th>Count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Species with TFBS at Position  Count\n",
       "0  (MEMB002A, MEMB002B, MEMB002C, MEMB002D, MEMB0...     17"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def most_frequent_pattern(alignment_num):\n",
    "    \"\"\"\n",
    "    alignment_num: integer specifying which alignment to analyze\n",
    "    \n",
    "    Returns:\n",
    "    A single-row table with the TFBS combo with the LARGEST count in the sequence\n",
    "    \"\"\"\n",
    "    max_count = max(list(counts_at_positions_tables[alignment_num][\"Count\"]))\n",
    "    table = group_by_combo_tables[alignment_num]\n",
    "    return table[table[\"Count\"] == max_count]\n",
    "\n",
    "most_frequent_pattern(1) # <- In the second alignment, the most frequent combo occurs 52 times."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "freq_tables = []\n",
    "for i in range(len(tables)): # len(tables) represents the alignments\n",
    "    t = frequent_patterns(i)\n",
    "    t[\"Alignment File\"] = [i] * t.count()[0]\n",
    "    freq_tables.append(t)\n",
    "\n",
    "frequent_all_alignments = pd.concat(freq_tables).sort_values(by=['Count'], ascending=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "      <th>Count</th>\n",
       "      <th>Alignment File</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>(MEMB005D,)</td>\n",
       "      <td>138</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>(dkik,)</td>\n",
       "      <td>86</td>\n",
       "      <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>(MEMB007B,)</td>\n",
       "      <td>86</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>(MEMB002F,)</td>\n",
       "      <td>82</td>\n",
       "      <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>(MEMB002B,)</td>\n",
       "      <td>78</td>\n",
       "      <td>196</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Species with TFBS at Position  Count  Alignment File\n",
       "117                   (MEMB005D,)    138              34\n",
       "123                       (dkik,)     86             256\n",
       "119                   (MEMB007B,)     86               9\n",
       "41                    (MEMB002F,)     82             196\n",
       "8                     (MEMB002B,)     78             196"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frequent_all_alignments.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species with TFBS at Position</th>\n",
       "      <th>Count</th>\n",
       "      <th>Alignment File</th>\n",
       "      <th>Alignment File Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(MEMB005D,)</td>\n",
       "      <td>138</td>\n",
       "      <td>34</td>\n",
       "      <td>VT56450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(MEMB002F,)</td>\n",
       "      <td>26</td>\n",
       "      <td>34</td>\n",
       "      <td>VT56450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(MEMB002B, MEMB004A)</td>\n",
       "      <td>25</td>\n",
       "      <td>34</td>\n",
       "      <td>VT56450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(MEMB003B,)</td>\n",
       "      <td>25</td>\n",
       "      <td>34</td>\n",
       "      <td>VT56450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(MEMB003F,)</td>\n",
       "      <td>22</td>\n",
       "      <td>34</td>\n",
       "      <td>VT56450</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Species with TFBS at Position  Count  Alignment File Alignment File Name\n",
       "0                   (MEMB005D,)    138              34             VT56450\n",
       "1                   (MEMB002F,)     26              34             VT56450\n",
       "2          (MEMB002B, MEMB004A)     25              34             VT56450\n",
       "3                   (MEMB003B,)     25              34             VT56450\n",
       "4                   (MEMB003F,)     22              34             VT56450"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### Get alignment file name and merge with frequent_all_alignments\n",
    "\n",
    "align_file = []\n",
    "for i in files[:200]:\n",
    "     align_file.append(os.path.basename(i))\n",
    "\n",
    "align_file_dict = {}\n",
    "for i in range(200):\n",
    "       align_file_dict[i] = align_file[i]\n",
    "\n",
    "file_key =  pd.DataFrame.from_dict(align_file_dict, orient='index')\n",
    "file_key = file_key.reset_index()\n",
    "file_key.columns = [\"Alignment File\", \"Alignment File Name\"]\n",
    "\n",
    "file_key['Alignment File Name'] = file_key['Alignment File Name'].str.split('_').str[5]\n",
    "file_key['Alignment File Name'] = file_key['Alignment File Name'].str.replace('.fa', '')\n",
    "\n",
    "occurence = frequent_all_alignments.merge(file_key,  how='outer')\n",
    "\n",
    "#occurance.shape # 61955\n",
    "#frequent_all_alignments.shape # 61955\n",
    "\n",
    "occurence.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "occurence.to_csv('~/Desktop/occurance_zelda.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
